---
title: "differential abundance aldex2"
output: html_notebook
---

Goal: since there is a significant difference in the composition of clusters, find out who is driving the difference

Groups
* "late" vs. "mid"

* "perturbed M" vs. "perturbed G"

* "mid C" vs. "mid G"
* "mid C" vs. "mid M"
* "mid G" vs. "mid M"

* "late C" vs. "late G" 
* "late C" vs. "late M"
* "late G" vs. "late M"

* "late C" vs. the others 

Methods:
* Wilcox Sum Rank Test
* ALDEx2
* ANCOM-BC

Levels
* Phylum
* Class
* Order
* Family
* Genus
* Species
* OTU

# Install Packages

```{r}
# .cran_packages <- c("tidyverse", "cowplot", "picante", "vegan", "HMP", "dendextend", "rms", "devtools")
# .inst <- .cran_packages %in% installed.packages()
# if(any(!.inst)) {
#   install.packages(.cran_packages[!.inst])
# }
```

```{r}
# .bioc_packages <- c("phyloseq", "Biostrings", "DESeq2", "microbiome", "metagenomeSeq", "ALDEx2", "ANCOMBC")
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install(.bioc_packages, version = "3.14")
```

# Loading Required Packages

```{r}
library("ggplot2"); packageVersion("ggplot2") # plotting
```
```{r}
library("tidyverse"); packageVersion("tidyverse") # util              
```
```{r}
library("dplyr"); packageVersion("dplyr") # util
```
```{r}
library("phyloseq"); packageVersion("phyloseq") # data object                 
```
```{r}
library("ape"); packageVersion("ape") # taxa string manipulation
```
```{r}
library(microbiome); packageVersion("microbiome") # clr transformation
```
```{r}
library(dendextend); packageVersion("dendextend") # extend customization of dendograms
```
```{r}
library(vegan); packageVersion("vegan") # bray-curtis and other metrics
```

```{r}
# Manipulation and visualization of taxonomic data, particularly those from amplicon metagenomics research. (Foster, Sharpton, and GrÃ¼nwald 2017)
library(metacoder); packageVersion("metacoder")
```
```{r}
# defines taxonomic classes and functions to manipulate them. The goal is to use these classes as low level fundamental taxonomic classes that other R packages can build on and use. This is used by metacoder. 
library(taxa); packageVersion("taxa")
```
```{r}
library(ALDEx2); packageVersion("ALDEx2") # aldex2 test (compositional differential abundance)
```
# Initialization

```{r}
# ggplot2 package theme set
theme_set(theme_bw())

# control randomization
set.seed(42)

# save outputs
save_location = "../data/processed_FCF/"
```


# Load Phyloseq Data Object
```{r}
(ps <- readRDS(paste(save_location, "phyloseq.rds", sep="")))
# convert to taxmap format 
taxmap <- parse_phyloseq(ps)

# Get Relative Abundance transform values
ps_rel_abund = phyloseq::transform_sample_counts(ps, function(x){x / sum(x)})

# Get CLR transform values
ps_clr <- microbiome::transform(ps, "clr")
taxmap_clr <- parse_phyloseq(ps_clr)
```

# ALDEx2 Test

```{r}
# data frame to store test results
ALDEx2_results <- data.frame(matrix(nrow = 0, ncol = 4))

# ALDEx2_results <- read.csv2(paste(save_location, "ALDEx2_results.csv", sep=""),
#                            header=TRUE,
#                            sep=",")
# ALDEx2_results$X <- NULL
# ALDEx2_results
```


## Example Output
```{r}
psa <- phyloseq::subset_samples(ps, cluster == "initial" | cluster == "mid")

aldex2_da <- ALDEx2::aldex(
    data.frame(phyloseq::otu_table(psa)), 
    phyloseq::sample_data(psa)$cluster, 
    mc.samples = 1000,
    test="t", 
    effect = TRUE, 
    denom="iqlr")
```

## Helper Functions

```{r}
aldex2_on_cluster <- function(ps, test_id) {
  #Run ALDEx2
  aldex2_da <- ALDEx2::aldex(
    data.frame(phyloseq::otu_table(ps)), 
    phyloseq::sample_data(ps)$cluster, 
    mc.samples = 1000,
    test="t", 
    effect = TRUE, 
    denom="iqlr")
  
  #Note positive and negative treatment variable 
  positive <- substr(
    colnames(aldex2_da)[2], 
    start = 9, 
    stop = nchar(colnames(aldex2_da)[2])
  )
  
  negative <- substr(
    colnames(aldex2_da)[3], 
    start = 9, 
    stop = nchar(colnames(aldex2_da)[3])
  )
  
  #Plot effect sizes
  png(file=paste(save_location, "aldex2_differential_abundance/", 
                 test_id, ".png", sep=""))
  par(mfrow=c(1,2))
  ALDEx2::aldex.plot(aldex2_da, type="MA", test="wilcox")
  ALDEx2::aldex.plot(aldex2_da, type="MW", test="wilcox")
  dev.off()
  
  #Clean up presentation
  colnames(aldex2_da) <- paste("aldex2", colnames(aldex2_da), sep = "_")
  sig_aldex2 <- aldex2_da %>%
    rownames_to_column(var = "OTU") %>%
    dplyr::select(OTU, aldex2_diff.btw, aldex2_diff.win, 
                  aldex2_effect, aldex2_wi.ep, aldex2_wi.eBH)
  
  #Get significant names and values to data frame
  sig_aldex2 <- merge(
    x=sig_aldex2, 
    y=phyloseq::tax_table(ps), 
    by.x="OTU", by.y=0, all.x=TRUE)
  
  #Save treatment variable
  sig_aldex2$aldex2_positive <- positive
  sig_aldex2$aldex2_negative <- negative
  
  return(sig_aldex2)
}

aldex2_on_focused_cluster <- function(ps, test_id) {
  #Run ALDEx2
  aldex2_da <- ALDEx2::aldex(
    data.frame(phyloseq::otu_table(ps)), 
    phyloseq::sample_data(ps)$focused.cluster,
    mc.samples = 1000,
    test="t", 
    effect = TRUE, 
    denom="iqlr")
  
  #Note positive and negative treatment variable 
  positive <- substr(
    colnames(aldex2_da)[2], 
    start = 9, 
    stop = nchar(colnames(aldex2_da)[2])
  )
  
  negative <- substr(
    colnames(aldex2_da)[3], 
    start = 9, 
    stop = nchar(colnames(aldex2_da)[3])
  )
  
  #Plot effect sizes
  png(file=paste(save_location, "aldex2_differential_abundance/", 
                 test_id, ".png", sep=""))
  par(mfrow=c(1,2))
  ALDEx2::aldex.plot(aldex2_da, type="MA", test="wilcox")
  ALDEx2::aldex.plot(aldex2_da, type="MW", test="wilcox")
  dev.off()
  
  #Clean up presentation
  colnames(aldex2_da) <- paste("aldex2", colnames(aldex2_da), sep = "_")
  sig_aldex2 <- aldex2_da %>%
    rownames_to_column(var = "OTU") %>%
    dplyr::select(OTU, aldex2_diff.btw, aldex2_diff.win, 
                  aldex2_effect, aldex2_wi.ep, aldex2_wi.eBH)
  
  #Get significant names and values to data frame
  sig_aldex2 <- merge(
    x=sig_aldex2, 
    y=phyloseq::tax_table(ps), 
    by.x="OTU", by.y=0, all.x=TRUE)
  
  #Save treatment variable
  sig_aldex2$aldex2_positive <- positive
  sig_aldex2$aldex2_negative <- negative
  
  return(sig_aldex2)
}
```

## Drivers
* test at each taxrank

```{r}
rank_names(ps)
```

```{r}
drive_aldex2_on_cluster <- function(ps, test_id) {
  # data frame to store test results
  results <- data.frame(matrix(nrow = 0, ncol = 4))
  
  # prune empty taxa
  ps <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
  
  # perform once at OTU level
  rows <- aldex2_on_cluster(ps, paste(test_id, "OTU", sep="_"))
  if (nrow(rows) > 0) {
    rows$taxrank <- "OTU"
    results <- rbind(results, rows) 
  }
  print("OTU") # logging
  
  # aggregate at each taxrank
  for (taxrank in rev(rank_names(ps)[-1])) {
    ps.taxa <- tax_glom(ps, taxrank = taxrank, NArm = TRUE)
    # skip if less than 10 taxa
    if (nrow(tax_table(ps.taxa)) < 11) {
      next
    }
    tryCatch(
      {
        # get result of test
        rows <- aldex2_on_cluster(ps.taxa, paste(test_id, taxrank, sep="_"))
        if (nrow(rows) > 0) {
          rows$taxrank <- taxrank
          results <- rbind(results, rows) 
        }
        print(taxrank) # logging
      },
      error = function(cond) {
        message("TEST FAILED!")
        message(cond)
      }
    )
  }
  
  results$test_id <- test_id
  
  return(results)
}


drive_aldex2_on_focused_cluster <- function(ps, test_id) {
  # data frame to store test results
  results <- data.frame(matrix(nrow = 0, ncol = 4))
  
  # prune empty taxa
  ps <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
  
  # perform once at OTU level
  rows <- aldex2_on_focused_cluster(ps, paste(test_id, "OTU", sep="_"))
  if (nrow(rows) > 0) {
    rows$taxrank <- "OTU"
    results <- rbind(results, rows) 
  }
  print("OTU") # logging
  
  # aggregate at each taxrank
  for (taxrank in rev(rank_names(ps)[-1])) {
    ps.taxa <- tax_glom(ps, taxrank = taxrank, NArm = TRUE)
    # skip if less than 10 taxa
    if (nrow(tax_table(ps.taxa)) < 11) {
      next
    }
    tryCatch(
      {
        # get result of test
        rows <- aldex2_on_focused_cluster(ps.taxa, paste(test_id, taxrank, sep="_"))
        if (nrow(rows) > 0) {
          rows$taxrank <- taxrank
          results <- rbind(results, rows) 
        }
        print(taxrank) # logging
      },
      error = function(cond) {
        message("TEST FAILED!")
        message(cond)
      }
    )
  }
  
  results$test_id <- test_id
  
  return(results)
}
```

## Clusters

### Cluster -- Pair-wise Comparisons

```{r}
rows <- drive_aldex2_on_cluster(phyloseq::subset_samples(ps, cluster == "initial" | cluster == "mid"), "i_vs_m")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
```{r}
rows <- drive_aldex2_on_cluster(phyloseq::subset_samples(ps, cluster == "initial" | cluster == "late"), "i_vs_l")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
```{r}
rows <- drive_aldex2_on_cluster(phyloseq::subset_samples(ps, cluster == "late" | cluster == "mid"), "l_vs_m")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
### Focused Cluster -- Pair-wise Comparisons

```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "perturbed G" | focused.cluster == "perturbed M"), "pG_vs_pM")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```

```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "mid C" | focused.cluster == "mid M"), "mC_vs_mM")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "mid C" | focused.cluster == "mid G"), "mC_vs_mG")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "mid G" | focused.cluster == "mid M"), "mG_vs_mM")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```

```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "late M"), "lC_vs_lM")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "late G"), "lC_vs_lG")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late G" | focused.cluster == "late M"), "lG_vs_lM")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
## Focused Cluster -- Pair-wise Comparisons to sC

```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "perturbed M"), "lC_vs_pM")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "perturbed G"), "lC_vs_pG")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```

```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "mid C"), "lC_vs_mC")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "mid G"), "lC_vs_mG")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```
```{r}
rows <- drive_aldex2_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "mid M"), "lC_vs_mM")
ALDEx2_results <- rbind(ALDEx2_results, rows)
```



### Save to .csv

```{r}
ALDEx2_results
```

```{r}
# save to csv
write.csv(ALDEx2_results, paste(save_location, "ALDEx2_results.csv", sep=""), quote = FALSE)
```

```{r}
# save only significantly differentially abundant taxa
ALDEx2_results %>%
  filter(aldex2_wi.eBH < 0.05) %>%
  write.csv(paste(save_location, "ALDEx2_results_TRUE.csv", sep=""), quote = FALSE)
```

