---
title: "differential abundance ANCOM-BC"
output: html_notebook
---

Goal: since there is a significant difference in the composition of clusters, find out who is driving the difference

Groups
* "stable" vs. "recovering"

* "perturbed M" vs. "perturbed G"

* "recovering C" vs. "recovering G"
* "recovering C" vs. "recovering M"
* "recovering G" vs. "recovering M"

* "stable C" vs. "stable G" 
* "stable C" vs. "stable M"
* "stable G" vs. "stable M"

* "stable C" vs. the others 

Methods:
* Wilcox Sum Rank Test
* ALDEx2
* ANCOM-BC

Levels
* Phylum
* Class
* Order
* Family
* Genus
* Species
* OTU

# Install Packages

```{r}
# .cran_packages <- c("tidyverse", "cowplot", "picante", "vegan", "HMP", "dendextend", "rms", "devtools")
# .inst <- .cran_packages %in% installed.packages()
# if(any(!.inst)) {
#   install.packages(.cran_packages[!.inst])
# }
```

```{r}
# .bioc_packages <- c("phyloseq", "Biostrings", "DESeq2", "microbiome", "metagenomeSeq", "ALDEx2", "ANCOMBC")
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install(.bioc_packages, version = "3.14")
```

# Loading Required Packages

```{r}
library("ggplot2"); packageVersion("ggplot2") # plotting
```
```{r}
library("tidyverse"); packageVersion("tidyverse") # util              
```
```{r}
library("dplyr"); packageVersion("dplyr") # util
```
```{r}
library("phyloseq"); packageVersion("phyloseq") # data object                 
```
```{r}
library("ape"); packageVersion("ape") # taxa string manipulation
```
```{r}
library(microbiome); packageVersion("microbiome") # clr transformation
```
```{r}
library(dendextend); packageVersion("dendextend") # extend customization of dendograms
```
```{r}
library(vegan); packageVersion("vegan") # bray-curtis and other metrics
```

```{r}
# Manipulation and visualization of taxonomic data, particularly those from amplicon metagenomics research. (Foster, Sharpton, and GrÃ¼nwald 2017)
library(metacoder); packageVersion("metacoder")
```
```{r}
# defines taxonomic classes and functions to manipulate them. The goal is to use these classes as low level fundamental taxonomic classes that other R packages can build on and use. This is used by metacoder. 
library(taxa); packageVersion("taxa")
```

```{r}
library(ANCOMBC); packageVersion("ANCOMBC") # ancomm-bc test(compositional differential abundance)
```

# Initialization

```{r}
# ggplot2 package theme set
theme_set(theme_bw())

# control randomization
set.seed(42)

# save outputs
save_location = "../data/processed_FCF/"
```

# Load Phyloseq Data Object
```{r}
(ps <- readRDS(paste(save_location, "phyloseq.rds", sep="")))
# convert to taxmap format 
taxmap <- parse_phyloseq(ps)

# Get Relative Abundance transform values
ps_rel_abund = phyloseq::transform_sample_counts(ps, function(x){x / sum(x)})

# Get CLR transform values
ps_clr <- microbiome::transform(ps, "clr")
taxmap_clr <- parse_phyloseq(ps_clr)
```

# ANCOM-BC Test

```{r}
# data frame to store test results
ANCOM_results <- data.frame(matrix(nrow = 0, ncol = 4))

# ANCOM_results <- read.csv2(paste(save_location, "ANCOM_results.csv", sep=""), 
#                            header=TRUE,
#                            sep=",")
# ANCOM_results$X <- NULL
# ANCOM_results
```

## Helper Functions

```{r}
psa <- phyloseq::subset_samples(ps, cluster == "early" | cluster == "recovering")
```

```{r}
ancom_on_cluster <- function(ps, test_id) {
  #Run ANCOM
  out <- ancombc(
    phyloseq = ps, 
    formula = "cluster", 
    p_adj_method = "holm", #default
    zero_cut = 0.90, #default
    lib_cut = 0, #default
    group = "cluster", 
    struc_zero = TRUE, 
    neg_lb = FALSE, #default
    tol = 1e-5, #default
    max_iter = 100, #default
    conserve = TRUE, 
    alpha = 0.05, #default
    global = TRUE
  )
  
  #Note positive and negative treatment variable 
  positive <- substr(
    names(out$res$beta)[1], 
    start = 8, 
    stop = nchar(names(out$res$beta)[1])
  )
  x1 <- unique(sample_data(psa)$cluster) 
  x2 <- x1[!x1 == positive]
  negative <- x2[1]
  
  
  
  #Clean up presentation
  names(out$res$beta)[1] <- "ancom_beta"
  names(out$res$se)[1] <- "ancom_se"
  names(out$res$W)[1] <- "ancom_W"
  names(out$res$p_val)[1] <- "ancom_p_val"
  names(out$res$q_val)[1] <- "ancom_q_val"
  names(out$res$diff_abn)[1] <- "ancom_diff_abn"
  
  #Merge all data frames
  sig_ancom <- out$res$beta %>%
      rownames_to_column(var = "OTU")
  sig_ancom <- merge(x=sig_ancom, y=out$res$se, by.x="OTU", by.y=0, all.x=TRUE)
  sig_ancom <- merge(x=sig_ancom, y=out$res$W, by.x="OTU", by.y=0, all.x=TRUE)
  sig_ancom <- merge(x=sig_ancom, y=out$res$p_val, by.x="OTU", by.y=0, all.x=TRUE)
  sig_ancom <- merge(x=sig_ancom, y=out$res$q_val, by.x="OTU", by.y=0, all.x=TRUE)
  sig_ancom <- merge(x=sig_ancom, y=out$res$diff_abn, by.x="OTU", by.y=0, all.x=TRUE)
  
  #Get significant names and values to data frame
  sig_ancom <- merge(x=sig_ancom, y=phyloseq::tax_table(ps), by.x="OTU", by.y=0, all.x=TRUE)
  
  #Save treatment variable
  sig_ancom$ancom_positive <- positive
  sig_ancom$ancom_negative <- negative
  
  return(sig_ancom)
}
  

ancom_on_focused_cluster <- function(ps, test_id) {
  #Run ANCOM
  out <- ancombc(
    phyloseq = ps, 
    formula = "focused.cluster", 
    p_adj_method = "holm", #default
    zero_cut = 0.90, #default
    lib_cut = 0, #default
    group = "focused.cluster", 
    struc_zero = TRUE, 
    neg_lb = FALSE, #default
    tol = 1e-5, #default
    max_iter = 100, #default
    conserve = TRUE, 
    alpha = 0.05, #default
    global = TRUE
  )
  
  #Note positive and negative treatment variable 
  positive <- substr(
    names(out$res$beta)[1], 
    start = 8, 
    stop = nchar(names(out$res$beta)[1])
  )
  x1 <- unique(sample_data(psa)$focused.cluster) 
  x2 <- x1[!x1 == positive]
  negative <- x2[1]

  #Clean up presentation
  names(out$res$beta)[1] <- "ancom_beta"
  names(out$res$se)[1] <- "ancom_se"
  names(out$res$W)[1] <- "ancom_W"
  names(out$res$p_val)[1] <- "ancom_p_val"
  names(out$res$q_val)[1] <- "ancom_q_val"
  names(out$res$diff_abn)[1] <- "ancom_diff_abn"
  
  #Merge all data frames
  sig_ancom <- out$res$beta %>%
      rownames_to_column(var = "OTU")
  sig_ancom <- merge(x=sig_ancom, y=out$res$se, by.x="OTU", by.y=0, all.x=TRUE)
  sig_ancom <- merge(x=sig_ancom, y=out$res$W, by.x="OTU", by.y=0, all.x=TRUE)
  sig_ancom <- merge(x=sig_ancom, y=out$res$p_val, by.x="OTU", by.y=0, all.x=TRUE)
  sig_ancom <- merge(x=sig_ancom, y=out$res$q_val, by.x="OTU", by.y=0, all.x=TRUE)
  sig_ancom <- merge(x=sig_ancom, y=out$res$diff_abn, by.x="OTU", by.y=0, all.x=TRUE)
  
  #Get significant names and values to data frame
  sig_ancom <- merge(x=sig_ancom, y=phyloseq::tax_table(ps), by.x="OTU", by.y=0, all.x=TRUE)
  
  #Save treatment variable
  sig_ancom$ancom_positive <- positive
  sig_ancom$ancom_negative <- negative
  
  return(sig_ancom)
}
```

## Drivers
* test at each taxrank

```{r}
rank_names(ps)
```

```{r}
drive_ancom_on_cluster <- function(ps, test_id) {
  # data frame to store test results
  results <- data.frame(matrix(nrow = 0, ncol = 4))
  
  # prune empty taxa
  ps <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
  
  # perform once at OTU level
  rows <- ancom_on_cluster(ps, paste(test_id, "OTU", sep="_"))
  if (nrow(rows) > 0) {
    rows$taxrank <- "OTU"
    results <- rbind(results, rows) 
  }
  print("OTU") # logging
  
  # aggregate at each taxrank
  for (taxrank in rev(rank_names(ps)[-1])) {
    ps.taxa <- tax_glom(ps, taxrank = taxrank, NArm = TRUE)
    # skip if less than 10 taxa
    if (nrow(tax_table(ps.taxa)) < 11) {
      next
    }
    tryCatch(
      {
        # get result of test
        rows <- ancom_on_cluster(ps.taxa, paste(test_id, taxrank, sep="_"))
        if (nrow(rows) > 0) {
          rows$taxrank <- taxrank
          results <- rbind(results, rows) 
        }
        print(taxrank) # logging
      },
      error = function(cond) {
        message("TEST FAILED!")
        message(cond)
      }
    )
  }
  
  results$test_id <- test_id
  
  return(results)
}


drive_ancom_on_focused_cluster <- function(ps, test_id) {
  # data frame to store test results
  results <- data.frame(matrix(nrow = 0, ncol = 4))
  
  # prune empty taxa
  ps <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
  
  # perform once at OTU level
  rows <- ancom_on_focused_cluster(ps, paste(test_id, "OTU", sep="_"))
  if (nrow(rows) > 0) {
    rows$taxrank <- "OTU"
    results <- rbind(results, rows) 
  }
  print("OTU") # logging
  
  # aggregate at each taxrank
  for (taxrank in rev(rank_names(ps)[-1])) {
    ps.taxa <- tax_glom(ps, taxrank = taxrank, NArm = TRUE)
    # skip if less than 10 taxa
    if (nrow(tax_table(ps.taxa)) < 11) {
      next
    }
    tryCatch(
      {
        # get result of test
        rows <- ancom_on_focused_cluster(
          ps.taxa, paste(test_id, taxrank, sep="_"))
        if (nrow(rows) > 0) {
          rows$taxrank <- taxrank
          results <- rbind(results, rows) 
        }
        print(taxrank) # logging
      },
      error = function(cond) {
        message("TEST FAILED!")
        message(cond)
      }
    )
  }
  
  results$test_id <- test_id
  
  return(results)
}
```

## Clusters

### Cluster -- Pair-wise Comparisons

```{r}
rows <- drive_ancom_on_cluster(phyloseq::subset_samples(ps, cluster == "early" | cluster == "recovering"), "early_vs_recovering")
ANCOM_results <- rbind(ANCOM_results, rows)
```
```{r}
rows <- drive_ancom_on_cluster(phyloseq::subset_samples(ps, cluster == "early" | cluster == "stable"), "early_vs_stable")
ANCOM_results <- rbind(ANCOM_results, rows)
```
```{r}
rows <- drive_ancom_on_cluster(phyloseq::subset_samples(ps, cluster == "stable" | cluster == "recovering"), "stable_vs_recovering")
ANCOM_results <- rbind(ANCOM_results, rows)
```
### Focused Cluster -- Pair-wise Comparisons

```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "perturbed G" | focused.cluster == "perturbed M"), "pG_vs_pM")
ANCOM_results <- rbind(ANCOM_results, rows)
```

```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "recovering C" | focused.cluster == "recovering M"), "rC_vs_rM")
ANCOM_results <- rbind(ANCOM_results, rows)
```
```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "recovering C" | focused.cluster == "recovering G"), "rC_vs_rG")
ANCOM_results <- rbind(ANCOM_results, rows)
```
```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "recovering G" | focused.cluster == "recovering M"), "rG_vs_rM")
ANCOM_results <- rbind(ANCOM_results, rows)
```

```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "stable C" | focused.cluster == "stable M"), "sC_vs_sM")
ANCOM_results <- rbind(ANCOM_results, rows)
```
```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "stable C" | focused.cluster == "stable G"), "sC_vs_sG")
ANCOM_results <- rbind(ANCOM_results, rows)
```
```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "stable G" | focused.cluster == "stable M"), "sG_vs_sM")
ANCOM_results <- rbind(ANCOM_results, rows)
```

## Focused Cluster -- Pair-wise Comparisons to sC

```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "stable C" | focused.cluster == "perturbed M"), "sC_vs_pM")
ANCOM_results <- rbind(ANCOM_results, rows)
```
```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "stable C" | focused.cluster == "perturbed G"), "sC_vs_pG")
ANCOM_results <- rbind(ANCOM_results, rows)
```

```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "stable C" | focused.cluster == "recovering C"), "sC_vs_rC")
ANCOM_results <- rbind(ANCOM_results, rows)
```
```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "stable C" | focused.cluster == "recovering G"), "sC_vs_rG")
ANCOM_results <- rbind(ANCOM_results, rows)
```
```{r}
rows <- drive_ancom_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "stable C" | focused.cluster == "recovering M"), "sC_vs_rM")
ANCOM_results <- rbind(ANCOM_results, rows)
```

### Save to .csv

```{r}
ANCOM_results
```

```{r}
# save to csv
write.csv(ANCOM_results, paste(save_location, "ANCOM_results.csv", sep=""), quote = FALSE)
```

```{r}
# save only significantly differentially abundant taxa
ANCOM_results %>%
  filter(ancom_diff_abn) %>%
  write.csv(paste(save_location, "ANCOM_results_TRUE.csv", sep=""), quote = FALSE)
```

