---
title: "Multivariate Comparison"
output: html_notebook
author: Gati Aher
date: 02/12/2022
---

Goal: compare clusters

Multivariate Comparison
* perMANOVA
* ANOSIM

Levels
* Phylum
* Class
* Order
* Family
* Genus
* Species
* OTU-level

# Install Packages

```{r}
# .cran_packages <- c("tidyverse", "cowplot", "picante", "vegan", "HMP", "dendextend", "rms", "devtools")
# .inst <- .cran_packages %in% installed.packages()
# if(any(!.inst)) {
#   install.packages(.cran_packages[!.inst])
# }
```

```{r}
# .bioc_packages <- c("phyloseq", "Biostrings", "DESeq2", "microbiome", "metagenomeSeq", "ALDEx2")
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install(.bioc_packages, version = "3.14")
```

# Loading Required Packages

```{r}
library("ggplot2"); packageVersion("ggplot2") # plotting
```
```{r}
library("tidyverse"); packageVersion("tidyverse") # util              
```
```{r}
library("dplyr"); packageVersion("dplyr") # util
```
```{r}
library("phyloseq"); packageVersion("phyloseq") # data object                 
```
```{r}
library("ape"); packageVersion("ape") # taxa string manipulation
```
```{r}
library(microbiome); packageVersion("microbiome") # clr transformation
```
```{r}
library(dendextend); packageVersion("dendextend") # extend customization of dendograms
```
```{r}
library(vegan); packageVersion("vegan") # bray-curtis and other metrics
```


# Initialization

```{r}
# ggplot2 package theme set
theme_set(theme_bw())

# control randomization
set.seed(42)

# save outputs
save_location = "../data/processed_FCF/"
```


# Load Phyloseq Data Object
```{r}
(ps <- readRDS(paste(save_location, "phyloseq.rds", sep="")))

# Get Relative Abundance transform values
ps_rel_abund = phyloseq::transform_sample_counts(ps, function(x){x / sum(x)})

# Get CLR transform values
ps_clr <- microbiome::transform(ps, "clr")
```
```{r}
sample_data(ps)
```



# Explore Cluster Relative Abundance Patterns

```{r}
# barchart_levels <- c("early", "perturbed G", "perturbed M", "mid C", "mid G", "mid M", "late C", "late G", "late M")
barchart_levels <- c("perturbed G", "mid G", "late G", "perturbed M",  "mid M", "late M", "mid C",  "late C")
```


```{r}
# Plot class
phyloseq::plot_bar(ps, fill = "class") +
  geom_bar(aes(color = class, fill = class), stat = "identity", position = "stack") +
  labs(x = "", y = "Counts\n") +
  facet_wrap(~factor(focused.cluster, levels=barchart_levels), scales = "free", ncol=3) +
  theme(panel.background = element_blank(),
        axis.ticks.x=element_blank())
```
```{r}
# Plot class
phyloseq::plot_bar(ps_rel_abund, fill = "class") +
  geom_bar(aes(color = class, fill = class), stat = "identity", position = "stack") +
  labs(x = "", y = "Relative Abundance\n") +
  facet_wrap(~factor(focused.cluster, levels=barchart_levels), scales = "free", ncol=3) +
  theme(panel.background = element_blank(),
        axis.ticks.x=element_blank())
```

```{r}
# Plot phylum
phyloseq::plot_bar(ps_rel_abund, fill = "phylum") +
  geom_bar(aes(color = phylum, fill = phylum), stat = "identity", position = "stack") +
  labs(x = "", y = "Relative Abundance\n") +
  facet_wrap(~factor(focused.cluster, levels=barchart_levels), scales = "free", ncol=3) +
  theme(panel.background = element_blank(),
        axis.ticks.x=element_blank())
```

# Visualize PCA

```{r}
plot_clr_pca <- function(ps, taxrank="none", color="cluster") {
  #PCA via phyloseq
  if (taxrank == "none") {
    ord_clr <- ps %>%
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  } else {
    ord_clr <- ps %>%
      tax_glom(taxrank = taxrank, NArm = TRUE) %>% 
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  }
  
  #Scale axes and plot ordination
  clr1 <- ord_clr$CA$eig[1] / sum(ord_clr$CA$eig)
  clr2 <- ord_clr$CA$eig[2] / sum(ord_clr$CA$eig)
  phyloseq::plot_ordination(ps, ord_clr, type="samples", color=color) +
    geom_point(size = 2) +
    coord_fixed(clr2 / clr1) +
    stat_ellipse(aes(group = carbon), linetype = 2)
  
  # #Plot scree plot
  # phyloseq::plot_scree(ord_clr) + 
  #   geom_bar(stat="identity", fill = "blue") +
  #   labs(x = "\nAxis", y = "Proportion of Variance\n")
}
```

## OTU Level

```{r}
plot_clr_pca(ps, "none", "cluster")
plot_clr_pca(ps, "none", "focused.cluster")
```

## Species Level

```{r}
plot_clr_pca(ps, "species", "cluster")
plot_clr_pca(ps, "species", "focused.cluster")
```

## Genus Level

```{r}
plot_clr_pca(ps, "genus", "cluster")
plot_clr_pca(ps, "genus", "focused.cluster")
```

## Family Level

```{r}
plot_clr_pca(ps, "family", "cluster")
plot_clr_pca(ps, "family", "focused.cluster")
```


## Order Level

```{r}
plot_clr_pca(ps, "order", "cluster")
plot_clr_pca(ps, "order", "focused.cluster")
```

# Visualize PERMANOVA


## Represent Dispersion Plot

```{r}
# NOTE: has to be same as levels for betadispr groups
focused.cluster_levels <- c("late C", "late G", "late M",
                            "mid C", "mid G", "mid M",
                            "perturbed G", "perturbed M")

focused.cluster_colors <- c("springgreen4", "steelblue4", "sienna4",
                            "springgreen3", "steelblue3", "sienna3",
                            "steelblue1", "sienna1")
names(focused.cluster_colors) <- focused.cluster_levels

focused.cluster_shapes <- c(15, 15, 15, 
                            12, 12, 12,
                            0, 0)
names(focused.cluster_shapes) <- focused.cluster_levels
```


```{r}
#Generate distance matrix
clr_dist_matrix <- phyloseq::distance(ps_clr, method = "euclidean")

all_test <- vegan::adonis(clr_dist_matrix ~ phyloseq::sample_data(ps_clr)$focused.cluster, permutations=10000)
print(all_test)

#get dispersions
dispr <- vegan::betadisper(clr_dist_matrix,
                           group = phyloseq::sample_data(ps_clr)$focused.cluster)
dispr$group

# plot dispersions distances
plot(dispr, 
     main = "Ordination Centroids and Dispersion: Aitchison Distance", 
     hull=FALSE, ellipse=TRUE, label = TRUE, label.cex=0.60,
     pch = focused.cluster_shapes,
     col = focused.cluster_colors
     )

# plot dispersions distances
plot(dispr, 
     main = "Ordination Centroids and Dispersion: Aitchison Distance", 
     hull=FALSE, ellipse=TRUE, label = FALSE, label.cex=0.60,
     pch = focused.cluster_shapes,
     col = focused.cluster_colors
     )
```


```{r}
# plot dispersions distances
plot(dispr, 
     main = "Ordination Centroids and Dispersion: Aitchison Distance", 
     hull=FALSE, ellipse=TRUE, label = FALSE, label.cex=0.60,
     conf = 0.68,
     pch = focused.cluster_shapes,
     col = focused.cluster_colors
     )
```


```{r}
# order factors
dispr_df <- data.frame(group = dispr$group, distances = dispr$distances)
dispr_df$group <- factor(dispr$group, 
                      levels = rev(focused.cluster_levels))

dispr_df %>%
  ggplot(aes(x=group, y=distances, colour=group, shape=group)) +
  geom_violin(trim = FALSE) +
  geom_jitter(position = position_jitter(0.2)) +
  scale_color_manual(values = focused.cluster_colors) +
  scale_shape_manual(values = focused.cluster_shapes) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


# PERMANOVA

```{r}
permutations=10000

# data frame to store test results
# "test_id", "R2", "p_value", "p_value_dispersion")
PERMANOVA_results <- data.frame(matrix(nrow = 0, ncol = 4))
```

## Helper Functions

```{r}
permanova_on_cluster <- function(ps, test_id) {
  #Get CLR transform values
  ps_clr <- microbiome::transform(ps, "clr")
  
  #Generate distance matrix
  clr_dist_matrix <- phyloseq::distance(ps_clr, method = "euclidean") 
  
  ###############
  # ADONIS test #
  ###############
  all_test <- vegan::adonis(clr_dist_matrix ~ phyloseq::sample_data(ps_clr)$cluster, permutations=10000)
  # print(all_test)
  
  #get dispersions
  dispr <- vegan::betadisper(clr_dist_matrix, phyloseq::sample_data(ps_clr)$cluster)
  dispr
  
  # # plot dispersions distances
  # plot(dispr, main = "Ordination Centroids and Dispersion Labeled: Aitchison Distance", sub ="")
  # 
  # # plot dispersions boxplot
  # boxplot(dispr, main = "", xlab = "")
  
  ################################################################
  # Permutation test for homogeneity of multivariate dispersions #
  ################################################################
  pd <- permutest(dispr)
  # print(pd)
  
  R2 <- all_test[["aov.tab"]][["R2"]][1] 
  p_value <- all_test[["aov.tab"]][["Pr(>F)"]][1]
  p_value_dispersion <- pd[["tab"]][["Pr(>F)"]][1]
  res = data.frame(test_id, R2, p_value, p_value_dispersion)
  
  return(res)
}

permanova_on_focused_cluster <- function(ps, test_id) {
  # Get CLR transform values
  ps_clr <- microbiome::transform(ps, "clr")
  
  #Generate distance matrix
  clr_dist_matrix <- phyloseq::distance(ps_clr, method = "euclidean") 
  
  ###############
  # ADONIS test #
  ###############
  all_test <- vegan::adonis(clr_dist_matrix ~ phyloseq::sample_data(ps_clr)$focused.cluster, permutations=10000)
  # print(all_test)
  
  #get dispersions
  dispr <- vegan::betadisper(clr_dist_matrix, phyloseq::sample_data(ps_clr)$focused.cluster)
  dispr
  
  # # plot dispersions distances
  # plot(dispr, main = "Ordination Centroids and Dispersion Labeled: Aitchison Distance", sub ="")
  # 
  # # plot dispersions boxplot
  # boxplot(dispr, main = "", xlab = "")
  
  ################################################################
  # Permutation test for homogeneity of multivariate dispersions #
  ################################################################
  pd <- permutest(dispr)
  # print(pd)
  
  R2 <- all_test[["aov.tab"]][["R2"]][1] 
  p_value <- all_test[["aov.tab"]][["Pr(>F)"]][1]
  p_value_dispersion <- pd[["tab"]][["Pr(>F)"]][1]
  res = data.frame(test_id, R2, p_value, p_value_dispersion)
  
  return(res)
}
```

## Drivers
* test at each taxrank

```{r}
rank_names(ps)
```

```{r}
drive_permanova_on_cluster <- function(ps, test_id) {
  # data frame to store test results
  results <- data.frame(matrix(nrow = 0, ncol = 4))
  
  # prune empty taxa
  ps <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
  
  # perform once at OTU level
  row <- permanova_on_cluster(ps, paste(test_id, "OTU-level", sep="_"))
  results <- rbind(results, row)
  
  # aggregate at each taxrank
  for (taxrank in rev(rank_names(ps)[-1])) {
    ps.taxa <- tax_glom(ps, taxrank = taxrank, NArm = TRUE)
    # get result of test
    row <- permanova_on_cluster(ps.taxa, paste(test_id, taxrank, sep="_"))
    results <- rbind(results, row)
  }
  
  return(results)
}


drive_permanova_on_focused_cluster <- function(ps, test_id) {
  # data frame to store test results
  results <- data.frame(matrix(nrow = 0, ncol = 4))
  
  # prune empty taxa
  ps <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
  
  # perform once at OTU level
  row <- permanova_on_focused_cluster(ps, paste(test_id, "OTU-level", sep="_"))
  results <- rbind(results, row)
  
  # aggregate at each taxrank
  for (taxrank in rev(rank_names(ps)[-1])) {
    ps.taxa <- tax_glom(ps, taxrank = taxrank, NArm = TRUE)
    # get result of test
    row <- permanova_on_focused_cluster(ps.taxa, paste(test_id, taxrank, sep="_"))
    results <- rbind(results, row)
  }
  
  return(results)
}
```

## Cluster

```{r}
rows <- drive_permanova_on_cluster(ps, "all_cluster")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

### Cluster -- Pair-wise Comparisons

```{r}
rows <- drive_permanova_on_cluster(phyloseq::subset_samples(ps, cluster == "perturbed" | cluster == "mid"), "p_vs_m")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_cluster(phyloseq::subset_samples(ps, cluster == "perturbed" | cluster == "late"), "p_vs_l")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_cluster(phyloseq::subset_samples(ps, cluster == "late" | cluster == "mid"), "l_vs_m")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

## Focused Cluster

```{r}
rows <- drive_permanova_on_focused_cluster(ps, "all_focused.cluster")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, cluster == "perturbed"), "p_focused.cluster")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, cluster == "mid"), "m_focused.cluster")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, cluster == "late"), "l_focused.cluster")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

### Focused Cluster -- Pair-wise Comparisons

```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "perturbed G" | focused.cluster == "perturbed M"), "pG_vs_pM")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "mid C" | focused.cluster == "mid M"), "mC_vs_mM")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "mid C" | focused.cluster == "mid G"), "mC_vs_mG")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "mid G" | focused.cluster == "mid M"), "mG_vs_mM")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "late M"), "lC_vs_lM")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "late G"), "lC_vs_lG")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late G" | focused.cluster == "late M"), "lG_vs_lM")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

## Focused Cluster -- Pair-wise Comparisons to late C

```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "perturbed G"), "lC_vs_pG")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "perturbed M"), "lC_vs_pM")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "mid C"), "lC_vs_mC")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "mid G"), "lC_vs_mG")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```
```{r}
rows <- drive_permanova_on_focused_cluster(phyloseq::subset_samples(ps, focused.cluster == "late C" | focused.cluster == "mid M"), "lC_vs_mM")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

## Multiple Corrections

```{r}
PERMANOVA_results$adjusted_p_value <- sapply(PERMANOVA_results$p_value, p.adjust, method="BH")
```

```{r}
PERMANOVA_results$significant <- (PERMANOVA_results$adjusted_p_value < 0.05) & (PERMANOVA_results$p_value_dispersion > 0.05)
```

```{r}
PERMANOVA_results
```

### Save to .csv

```{r}
# save to csv
write.csv(PERMANOVA_results, paste(save_location, "PERMANOVA_results.csv", sep=""), quote = FALSE)
```

```{r}
# save significantly TRUE values to csv
write.csv(PERMANOVA_results[which(PERMANOVA_results$significant), ], paste(save_location, "PERMANOVA_results_TRUE.csv", sep=""), quote = FALSE)
```

