---
title: "sanity_ordination"
output: html_notebook
---

Goal: ordination for sanity check

| plots | loading |
| --------- | --------- |
| PCA, focused.cluster | all, colored by DA, rho |
| PCA, focused.cluster | DA, rho |
| PCA, focused.cluster | min abundance in any sample of 1% (rare OTUs excluded)  |
| PCA, focused.cluster | max abundance in any sample of 2% (common OTUs excluded) |
| PCA, focused.cluster | all |

# Loading Required Packages

```{r}
library("ggplot2"); packageVersion("ggplot2") # plotting
```
```{r}
library("tidyverse"); packageVersion("tidyverse") # util              
```
```{r}
library("dplyr"); packageVersion("dplyr") # util
```
```{r}
library("phyloseq"); packageVersion("phyloseq") # data object                 
```
```{r}
library("ape"); packageVersion("ape") # taxa string manipulation
```
```{r}
library(microbiome); packageVersion("microbiome") # clr transformation
```

```{r}
library(dendextend); packageVersion("dendextend") # extend customization of dendograms
```

```{r}
library(vegan); packageVersion("vegan") # bray-curtis and other metrics
```


# Initialization

```{r}
# ggplot2 package theme set
theme_set(theme_bw())

# control randomization
set.seed(42)

# save outputs
raw_location = "../data/raw/FCF/"
save_location = "../data/processed_FCF_53OTUs//"
```

# Dataset Prep

## Unflitered PS Object


```{r}
otu_mat <- as.matrix(read.table(file = paste(raw_location, 'processed_OTU_counts.tsv', sep = ""), sep = '\t', header = TRUE, row.names = 1))
tax_mat <- as.matrix(read.table(file = paste(raw_location, 'processed_taxonomy_table.tsv', sep = ""), sep = '\t', header = TRUE, row.names = 1))
sample_meta <- read.table(file = paste(raw_location, 'processed_sample_metadata.tsv', sep = ""), header = TRUE, sep = '\t', row.names = 1)
sample_meta$day <- factor(sample_meta$day, levels=c("3", "5", "7", "10", "20"))
sample_meta$transfer <- factor(sample_meta$transfe, levels=c("0", "1", "2", "3"))
# cluster
sample_meta$cluster <- factor(sample_meta$cluster, levels=c("early", "mid", "late", "perturbed"))
sample_meta$focused.cluster <- factor(sample_meta$focused.cluster, levels=c("early", "mid C", "late C", "perturbed G", "mid G", "late G", "perturbed M", "mid M", "late M"))

# get phylogenetic tree
tree <- ape::read.tree(paste(raw_location, "MUSCLE_alignment_ML_tree_all_OTUs", sep=""))

class(otu_mat)
class(tax_mat)
class(sample_meta)

OTU = otu_table(otu_mat, taxa_are_rows = TRUE)
TAX = tax_table(tax_mat)
unflitered_ps = phyloseq(OTU, TAX, tree)
sample_data(unflitered_ps) <- sample_meta
unflitered_ps
```
# Filtering Curve

```{r}
filtering_curve_MIN_PERCENT_IN_ANY_SAMPLE <- function(unfiltered_ps, MIN_PERCENT_IN_ANY_SAMPLE_vals = c(0, 0.001)) {
  # Return table of how many OTUs survive at each filtering threshold
  res = c()
  per = c()
  for (MIN_PERCENT_IN_ANY_SAMPLE in MIN_PERCENT_IN_ANY_SAMPLE_vals) {
    ps_rel_abund <- phyloseq::transform_sample_counts(unflitered_ps, function(x){x / sum(x)})
    res <- c(res, sum(rowSums( otu_table(ps_rel_abund) > (MIN_PERCENT_IN_ANY_SAMPLE / 100)) > 0L))
    per <- c(per, MIN_PERCENT_IN_ANY_SAMPLE)
  }
  names(res) <- per
  return(res)
}
```


```{r}
filtering_curve <- data.frame(
  filtering_curve_MIN_PERCENT_IN_ANY_SAMPLE(
    unfiltered_ps, 
    MIN_PERCENT_IN_ANY_SAMPLE_vals = c(0, 
                                       0.001,
                                       0.005,
                                       0.01,
                                       0.05,
                                       0.1,
                                       0.5, 
                                       1, 
                                       2,
                                       3)))

colnames(filtering_curve) <- c("OTUs")
filtering_curve
```

# Aesthetics

```{r}
# day
day_shade_colors <- c(
    "3" = "#B2E5EA", # lightest
    "5" = "#00BAE7",
    "7" = "#0E6FA0",
    "10" = "#012652",
    "20" = "#01091C" # darkest
  )

# group
group_shapes_cell <- c(
    "C0C" = 4,
    "1C" = 10,
    "2C" = 16,
    "2G" = 9,
    "3G" = 18,
    "2M" = 12,
    "3M" = 15
)

group_shapes_all <- c(
    "C0C" = 4,
    "1C" = 10,
    "2C" = 16,
    "1G" = 5,
    "2G" = 9,
    "3G" = 18,
    "1M" = 0,
    "2M" = 12,
    "3M" = 15
)

# cluster
cluster_colors <- c("perturbed" = "gray50", # early
                   "early" = "blue", # mid
                   "mid" = "green", # late
                   "late" = "red") # perturbed

# focused cluster
focused.cluster_levels <- c("early", "mid C", "late C", 
                              "perturbed G", "mid G", "late G", 
                              "perturbed M", "mid M", "late M")
  
focused.cluster_colors <- c("gray50", "springgreen3", "springgreen4",
                              "steelblue1", "steelblue3", "steelblue4",
                              "sienna1", "sienna3", "sienna4")
names(focused.cluster_colors) <- focused.cluster_levels
  
focused.cluster_shapes <- c(1, 10, 16,
                              1, 10, 16,
                              1, 10, 16)
names(focused.cluster_shapes) <- focused.cluster_levels
```


```{r}
plot_clr_pca <- function(ps, taxrank="none", color="cluster") {
  #PCA via phyloseq
  if (taxrank == "none") {
    ord_clr <- ps %>%
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  } else {
    ord_clr <- ps %>%
      tax_glom(taxrank = taxrank, NArm = TRUE) %>% 
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  }
  
  #Scale axes and plot ordination
  clr1 <- ord_clr$CA$eig[1] / sum(ord_clr$CA$eig)
  clr2 <- ord_clr$CA$eig[2] / sum(ord_clr$CA$eig)
  
  # samples colored by color
  phyloseq::plot_ordination(ps, ord_clr, type="samples", color=color) +
    geom_point(size = 2) +
    coord_fixed(clr2 / clr1) +
    stat_ellipse(aes(group = carbon), linetype = 2)
  
  # #Plot scree plot
  # phyloseq::plot_scree(ord_clr) + 
  #   geom_bar(stat="identity", fill = "blue") +
  #   labs(x = "\nAxis", y = "Proportion of Variance\n")
}


plot_taxa_pca <- function(ps, taxrank="none") {
  #PCA via phyloseq
  if (taxrank == "none") {
    ord_clr <- ps %>%
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  } else {
    ord_clr <- ps %>%
      tax_glom(taxrank = taxrank, NArm = TRUE) %>% 
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  }
  
  #Scale axes and plot ordination
  clr1 <- ord_clr$CA$eig[1] / sum(ord_clr$CA$eig)
  clr2 <- ord_clr$CA$eig[2] / sum(ord_clr$CA$eig)
  
  # type split (biplot)
  phyloseq::plot_ordination(ps, ord_clr, type="taxa", color="genus") +
    geom_point(size = 2) +
    coord_fixed(clr2 / clr1) #+
    # stat_ellipse(aes(group = carbon), linetype = 2)
  
  # #Plot scree plot
  # phyloseq::plot_scree(ord_clr) + 
  #   geom_bar(stat="identity", fill = "blue") +
  #   labs(x = "\nAxis", y = "Proportion of Variance\n")
}
```




```{r}
plot_clr_pca_focused_cluster <- function(ps, taxrank="none") {
  #PCA via phyloseq
  if (taxrank == "none") {
    ord_clr <- ps %>%
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  } else {
    ord_clr <- ps %>%
      tax_glom(taxrank = taxrank, NArm = TRUE) %>%
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  }
  
  # make data frame
  clr_pca.data <- data.frame(PC1=ord_clr$CA$u[,1], PC2=ord_clr$CA$u[,2])
  meta <- data.frame(phyloseq::sample_data(ps))
  clr_pca.data <- merge(clr_pca.data, meta, by="row.names")

  #Scale axes and plot ordination
  clr1 <- ord_clr$CA$eig[1] / sum(ord_clr$CA$eig)
  clr2 <- ord_clr$CA$eig[2] / sum(ord_clr$CA$eig)
  ggplot(clr_pca.data, aes(x = PC1, y = PC2, colour = focused.cluster, shape = focused.cluster)) +
    theme_classic() +
    geom_point(
      size = 3,
      alpha = 5/6
    ) +
    coord_fixed(clr2 / clr1) +
    # stat_ellipse(aes(group = focused.cluster), linetype = 2) +
    scale_color_manual(values = focused.cluster_colors) +
    scale_shape_manual(values = focused.cluster_shapes) +
    xlab(sprintf("PC 1 %s%%", round(100 * clr1, 2))) +
    ylab(sprintf("PC 2 %s%%", round(100 * clr2, 2))) +
    ggtitle(sprintf("PCA taxrank: %s", taxrank))

  # #Plot scree plot
  # phyloseq::plot_scree(ord_clr) +
  #   geom_bar(stat="identity", fill = "blue") +
  #   labs(x = "\nAxis", y = "Proportion of Variance\n")
}
```

```{r}
plot_PERMANOVA_pca <- function(ps) {
  #CLR transform
  ps_clr <- microbiome::transform(ps, "clr")
  
  #Generate distance matrix
  clr_dist_matrix <- phyloseq::distance(ps_clr, method = "euclidean")
  
  all_test <- vegan::adonis(clr_dist_matrix ~ phyloseq::sample_data(ps_clr)$focused.cluster, permutations=10000)
  print(all_test)
  
  #get dispersions
  dispr <- vegan::betadisper(clr_dist_matrix,
                             group = phyloseq::sample_data(ps_clr)$focused.cluster,
                             bias.adjust=TRUE)
  # dispr$group
  
  # plot dispersions distances
  plot(dispr, 
       main = "Ordination Centroids and Dispersion: Aitchison Distance", 
       hull=FALSE, ellipse=TRUE, label = TRUE, label.cex=0.60,
       pch = focused.cluster_shapes,
       col = focused.cluster_colors
       )
  
  # plot dispersions distances
  plot(dispr, 
       main = "Ordination Centroids and Dispersion: Aitchison Distance", 
       hull=FALSE, ellipse=TRUE, label = FALSE, label.cex=0.60,
       pch = focused.cluster_shapes,
       col = focused.cluster_colors
       )
  
  return(dispr)
}
```

```{r}
plot_pretty_clr_pca <- function(ps, taxrank="none") {
  #PCA via phyloseq
  if (taxrank == "none") {
    ord_clr <- ps %>%
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  } else {
    ord_clr <- ps %>%
      tax_glom(taxrank = taxrank, NArm = TRUE) %>% 
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  }
  
  # make data frame
  clr_pca.data <- data.frame(PC1=ord_clr$CA$u[,1], PC2=ord_clr$CA$u[,2])
  meta <- data.frame(phyloseq::sample_data(ps))
  clr_pca.data <- merge(clr_pca.data, meta, by="row.names")
  clr_pca.data$day <- as.factor(clr_pca.data$day)
  
  #Scale axes and plot ordination
  clr1 <- ord_clr$CA$eig[1] / sum(ord_clr$CA$eig)
  clr2 <- ord_clr$CA$eig[2] / sum(ord_clr$CA$eig)
  ggplot(clr_pca.data, aes(x = PC1, y = PC2)) + 
    theme_classic() + 
    geom_point(
      mapping = aes(colour = day, shape = carbon),
      size = 3,
      alpha = 5/6
    ) +
    coord_fixed(clr2 / clr1) +
    stat_ellipse(aes(group = carbon), linetype = 2) +
    scale_color_manual("day", values = day_shade_colors) +
    scale_shape_manual(values = c(
    "original" = 4,
    "cellulose" = 16,
    "glucose" = 18,
    "malate" = 15
    )) +
    xlab(sprintf("PC 1 %s%%", round(100 * clr1, 2))) +
    ylab(sprintf("PC 2 %s%%", round(100 * clr2, 2))) +
    ggtitle(sprintf("PCA taxrank: %s", taxrank))
  
  # #Plot scree plot
  # phyloseq::plot_scree(ord_clr) + 
  #   geom_bar(stat="identity", fill = "blue") +
  #   labs(x = "\nAxis", y = "Proportion of Variance\n")
}


plot_pretty_clr_pca_detailed <- function(ps, taxrank="none") {
  #PCA via phyloseq
  if (taxrank == "none") {
    ord_clr <- ps %>%
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  } else {
    ord_clr <- ps %>%
      tax_glom(taxrank = taxrank, NArm = TRUE) %>% 
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  }
  
  # make data frame
  clr_pca.data <- data.frame(PC1=ord_clr$CA$u[,1], PC2=ord_clr$CA$u[,2])
  meta <- data.frame(phyloseq::sample_data(ps))
  clr_pca.data <- merge(clr_pca.data, meta, by="row.names")
  clr_pca.data$day <- as.factor(clr_pca.data$day)
  
  #Scale axes and plot ordination
  clr1 <- ord_clr$CA$eig[1] / sum(ord_clr$CA$eig)
  clr2 <- ord_clr$CA$eig[2] / sum(ord_clr$CA$eig)
  ggplot(clr_pca.data, aes(x = PC1, y = PC2)) + 
    theme_classic() + 
    geom_point(
      mapping = aes(colour = day, shape = group),
      size = 3,
      alpha = 5/6
    ) +
    coord_fixed(clr2 / clr1) +
    stat_ellipse(aes(group = carbon), linetype = 2) +
    scale_color_manual("day", values = day_shade_colors) +
    scale_shape_manual("group", values = group_shapes_all) +
    xlab(sprintf("PC 1 %s%%", round(100 * clr1, 2))) +
    ylab(sprintf("PC 2 %s%%", round(100 * clr2, 2))) +
    ggtitle(sprintf("PCA taxrank: %s", taxrank))
  
  # #Plot scree plot
  # phyloseq::plot_scree(ord_clr) + 
  #   geom_bar(stat="identity", fill = "blue") +
  #   labs(x = "\nAxis", y = "Proportion of Variance\n")
}
```

## PCA on samples (minimum relative abundance in any sample of 1%) 
* rare OTUs excluded

```{r}
MIN_PERCENT_IN_ANY_SAMPLE <- 1
ps_rel_abund = phyloseq::transform_sample_counts(unflitered_ps, function(x){x / sum(x)})
(ps <- phyloseq::prune_taxa(rowSums(otu_table(ps_rel_abund) > MIN_PERCENT_IN_ANY_SAMPLE / 100) > 0L
                           , unflitered_ps))

data.frame(tax_table(ps))

# on all samples
plot_clr_pca(ps, taxrank="none", color="cluster")
plot_taxa_pca(ps)
plot_pretty_clr_pca(ps, taxrank = "none")
plot_pretty_clr_pca_detailed(ps, taxrank = "none")
plot_clr_pca_focused_cluster(ps, taxrank="none")
dispr <- plot_PERMANOVA_pca(ps)

# on cellulose samples
ps_cellulose <- phyloseq::subset_samples(ps, carbon == "cellulose" | carbon == "original") 
ps_cellulose <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps_cellulose) > 0, ps_cellulose)
plot_pretty_clr_pca_detailed(ps_cellulose, taxrank = "none")
```

## PCA on samples (maximum relative abundance in any sample of 2%) 
* common OTUs excluded

```{r}
MAX_PERCENT_IN_ANY_SAMPLE <- 2
ps_rel_abund = phyloseq::transform_sample_counts(unflitered_ps, function(x){x / sum(x)})
(ps <- phyloseq::prune_taxa(rowSums(otu_table(ps_rel_abund) < MAX_PERCENT_IN_ANY_SAMPLE / 100) == phyloseq::nsamples(ps)
                           , unflitered_ps))

data.frame(tax_table(ps))

# on all samples
plot_clr_pca(ps, taxrank="none", color="cluster")
plot_taxa_pca(ps)
plot_pretty_clr_pca(ps, taxrank = "none")
plot_pretty_clr_pca_detailed(ps, taxrank = "none")
plot_clr_pca_focused_cluster(ps, taxrank="none")
plot_PERMANOVA_pca(ps)

# on cellulose samples
ps_cellulose <- phyloseq::subset_samples(ps, carbon == "cellulose" | carbon == "original") 
ps_cellulose <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps_cellulose) > 0, ps_cellulose)
plot_pretty_clr_pca_detailed(ps_cellulose, taxrank = "none")
```

## PCA on samples (minimum relative abundance in any sample of 0.1%)
* 53 OTUs

```{r}
MIN_PERCENT_IN_ANY_SAMPLE <- 0.1
ps_rel_abund = phyloseq::transform_sample_counts(unflitered_ps, function(x){x / sum(x)})
(ps <- phyloseq::prune_taxa(rowSums(otu_table(ps_rel_abund) > MIN_PERCENT_IN_ANY_SAMPLE / 100) > 0L
                           , unflitered_ps))

data.frame(tax_table(ps))

# on all samples
plot_clr_pca(ps, taxrank="none", color="cluster")
plot_taxa_pca(ps)
plot_pretty_clr_pca(ps, taxrank = "none")
plot_pretty_clr_pca_detailed(ps, taxrank = "none")
plot_clr_pca_focused_cluster(ps, taxrank="none")
plot_PERMANOVA_pca(ps)

# on cellulose samples
ps_cellulose <- phyloseq::subset_samples(ps, carbon == "cellulose" | carbon == "original") 
ps_cellulose <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps_cellulose) > 0, ps_cellulose)
plot_pretty_clr_pca_detailed(ps_cellulose, taxrank = "none")
```

# Pretty Ward Dendogram

```{r}
get_ward <- function(ps, taxrank="none") {
  #PCA via phyloseq
  if (taxrank == "none") {
    ps_clr <- ps %>%
      microbiome::transform("clr")
  } else {
    ps_clr <- ps %>%
      tax_glom(taxrank = taxrank, NArm = TRUE) %>% 
      microbiome::transform("clr")
  }
  
  #Get Aitchison distance matrix
  ps_clr_otu <- data.frame(phyloseq::otu_table(ps_clr))
  ps_clr_otu <- t(ps_clr_otu)
  a_dist_clr <- vegan::vegdist(ps_clr_otu,  method = "euclidean")
  
  #Save as dendrogram
  ward <- as.dendrogram(hclust(a_dist_clr, method = "ward.D2"))
  
  #Provide color codes
  meta <- data.frame(phyloseq::sample_data(ps))
  colorCode <- c(C = "green", M = "orange", G = "blue")
  labels_colors(ward) <- colorCode[meta$series][order.dendrogram(ward)]
  
  #Plot color by code
  par(cex = 0.4)
  ward %>% plot(main = sprintf("Aitchison on Centered Log-Ratios \ntaxrank: %s", taxrank))
  
  #Plot color by cluster
  ward %>% set("labels_cex", 1) %>% set("labels_col", value = c(1, 2, 3), k=3) %>% 
     plot(main = sprintf("Aitchison: Color labels per cluster \ntaxrank: %s", taxrank))
  
  #Plot color by cluster
  ward %>% set("labels_cex", 1) %>% set("labels_col", value = c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10), k=9) %>% 
     plot(main = sprintf("Aitchison: Color labels per cluster \ntaxrank: %s", taxrank))
  
  #Fully annotated plot
  ward %>%
    # symbols
    set("leaves_pch", 
        group_shapes_all[meta$group][order.dendrogram(ward)]) %>%
    set("leaves_cex", c(2)) %>%
    set("leaves_col", 
        day_shade_colors[meta$day][order.dendrogram(ward)]) %>%
    # labels
    set("labels_col", 
        focused.cluster_colors[meta$focused.cluster][order.dendrogram(ward)]) %>%
    plot(main = sprintf("Aitchison on Centered Log-Ratios"))
  # colored bar
  colored_bars(colors = cluster_colors[meta$cluster], 
               dend = ward,
               rowLabels = "cluster",
               sort_by_labels_order = TRUE,
               y_shift = -6)
  
  return(ward)
}
```

```{r}
ps
ps_cellulose
```

```{r}
ward_cellulose_OTU <- get_ward(ps_cellulose, "none")
```

```{r}
ward <- get_ward(ps, taxrank = "none")
```






