---
title: "Multivariate Comparison"
output: html_notebook
author: Gati Aher
date: 02/12/2022
---

Goal: compare clusters

Multivariate Comparison
* perMANOVA
* ANOSIM

Levels
* Phylum
* Class
* Order
* Family
* Genus
* Species
* OTU-level

# Install Packages

```{r}
# .cran_packages <- c("tidyverse", "cowplot", "picante", "vegan", "HMP", "dendextend", "rms", "devtools")
# .inst <- .cran_packages %in% installed.packages()
# if(any(!.inst)) {
#   install.packages(.cran_packages[!.inst])
# }
```

```{r}
# .bioc_packages <- c("phyloseq", "Biostrings", "DESeq2", "microbiome", "metagenomeSeq", "ALDEx2")
# if (!requireNamespace("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
# BiocManager::install(.bioc_packages, version = "3.14")
```

# Loading Required Packages

```{r}
library("ggplot2"); packageVersion("ggplot2") # plotting
```
```{r}
library("tidyverse"); packageVersion("tidyverse") # util              
```
```{r}
library("dplyr"); packageVersion("dplyr") # util
```
```{r}
library("phyloseq"); packageVersion("phyloseq") # data object                 
```
```{r}
library("ape"); packageVersion("ape") # taxa string manipulation
```
```{r}
library(microbiome); packageVersion("microbiome") # clr transformation
```
```{r}
library(dendextend); packageVersion("dendextend") # extend customization of dendograms
```
```{r}
library(vegan); packageVersion("vegan") # bray-curtis and other metrics
```


# Initialization

```{r}
# ggplot2 package theme set
theme_set(theme_bw())

# control randomization
set.seed(42)

# save outputs
save_location = "../data/processed_carbon_53OTUs/"
```

# Aes

```{r}
# color - days
color_day = c(
    "3" = "#B2E5EA", # lightest
    "5" = "#00BAE7",
    "7" = "#0E6FA0",
    "10" = "#012652",
    "20" = "#01091C"
    )

# color - growth.phase
color_growth.phase = c(
  "early" = "black",
  "mid C:C" = "springgreen3",
  "late C:C" = "springgreen4",
  "perturbed G" = "steelblue1",
  "mid G:C" = "steelblue3",
  "late G:C" = "steelblue4",
  "perturbed M" = "sienna1",
  "mid M:C" = "sienna3",
  "late M:C" = "sienna4"
  )

# shape - series
shape_series = c(
  "C0C" = 4,
  "C" = 16,
  "G" = 15,
  "M" = 17
  )
# shape - carbon
shape_carbon = c(
  "original" = 4,
  "cellulose" = 16,
  "glucose" = 18,
  "malate" = 15
)
# shape - transfer
shape_transfer = c(
    "C0C" = 4,
    "1C" = 10,
    "2C" = 16,
    "1G" = 0,
    "2G" = 12,
    "3G" = 15,
    "1M" = 5,
    "2M" = 9,
    "3M" = 18
  )
# shape - growth phase
shape_growth.phase = c(
  "early" = 1,
  "mid C:C" = 16,
  "late C:C" = 16,
  "perturbed G" = 15,
  "mid G:C" = 15,
  "late G:C" = 15,
  "perturbed M" = 17,
  "mid M:C" = 17,
  "late M:C" = 17
  )

# color - genus
```


# Load Phyloseq Data Object
```{r}
(ps <- readRDS(paste(save_location, "phyloseq.rds", sep="")))

# Get Relative Abundance transform values
ps_rel_abund = phyloseq::transform_sample_counts(ps, function(x){x / sum(x)})

# Get CLR transform values
ps_clr <- microbiome::transform(ps, "clr")
```
```{r}
sample_data(ps)
```

# Explore Cluster Relative Abundance Patterns

```{r}
# Plot class
phyloseq::plot_bar(ps, fill = "class") +
  geom_bar(aes(color = class, fill = class), stat = "identity", position = "stack") +
  labs(x = "", y = "Counts\n") +
  facet_wrap(~factor(growth.phase,), scales = "free", ncol=3) +
  theme(panel.background = element_blank(),
        axis.ticks.x=element_blank())
```


```{r}
# Plot genus
phyloseq::plot_bar(ps_rel_abund, fill = "genus") +
  geom_bar(aes(color = genus, fill = genus), stat = "identity", position = "stack") +
  labs(x = "", y = "Relative Abundance\n") +
  facet_wrap(~factor(growth.phase), scales = "free", ncol=3) +
  theme(panel.background = element_blank(),
        axis.ticks.x=element_blank())

# Plot class
phyloseq::plot_bar(ps_rel_abund, fill = "class") +
  geom_bar(aes(color = class, fill = class), stat = "identity", position = "stack") +
  labs(x = "", y = "Relative Abundance\n") +
  facet_wrap(~factor(growth.phase), scales = "free", ncol=3) +
  theme(panel.background = element_blank(),
        axis.ticks.x=element_blank())

# Plot phylum
phyloseq::plot_bar(ps_rel_abund, fill = "phylum") +
  geom_bar(aes(color = phylum, fill = phylum), stat = "identity", position = "stack") +
  labs(x = "", y = "Relative Abundance\n") +
  facet_wrap(~factor(growth.phase), scales = "free", ncol=3) +
  theme(panel.background = element_blank(),
        axis.ticks.x=element_blank())
```
# Visualize PCA

```{r}
plot_clr_pca <- function(ps, taxrank="none", color="cluster") {
  #PCA via phyloseq
  if (taxrank == "none") {
    ord_clr <- ps %>%
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  } else {
    ord_clr <- ps %>%
      tax_glom(taxrank = taxrank, NArm = TRUE) %>% 
      microbiome::transform("clr") %>%
      phyloseq::ordinate("RDA")
  }
  
  #Scale axes and plot ordination
  clr1 <- ord_clr$CA$eig[1] / sum(ord_clr$CA$eig)
  clr2 <- ord_clr$CA$eig[2] / sum(ord_clr$CA$eig)
  phyloseq::plot_ordination(ps, ord_clr, type="samples", color=color) +
    geom_point(size = 2) +
    coord_fixed(clr2 / clr1) +
    stat_ellipse(aes(group = carbon), linetype = 2)
  
  # #Plot scree plot
  # phyloseq::plot_scree(ord_clr) + 
  #   geom_bar(stat="identity", fill = "blue") +
  #   labs(x = "\nAxis", y = "Proportion of Variance\n")
}
```

## OTU Level

```{r}
plot_clr_pca(ps, "none", "cluster")
plot_clr_pca(ps, "none", "growth.phase")
```
# Visualize PERMANOVA

## Represent Dispersion Plot

```{r}
levels(sample_data(ps)$growth.phase)
```

```{r}
#Generate distance matrix
clr_dist_matrix <- phyloseq::distance(ps_clr, method = "euclidean")

all_test <- vegan::adonis(clr_dist_matrix ~ phyloseq::sample_data(ps_clr)$growth.phase, permutations=10000)
print(all_test)

#get dispersions
dispr <- vegan::betadisper(clr_dist_matrix,
                           group = phyloseq::sample_data(ps_clr)$growth.phase,
                           bias.adjust=TRUE)
dispr$group

# plot dispersions distances
plot(dispr, 
     main = "Ordination Centroids and Dispersion: Aitchison Distance", 
     hull=FALSE, ellipse=TRUE, label = TRUE, label.cex=0.60,
     pch = shape_growth.phase,
     col = color_growth.phase
     )

# plot dispersions distances
plot(dispr, 
     main = "Ordination Centroids and Dispersion: Aitchison Distance", 
     hull=FALSE, ellipse=TRUE, label = FALSE, label.cex=0.60,
     pch = shape_growth.phase,
     col = color_growth.phase
     )
```


```{r}
# plot dispersions distances
plot(dispr, 
     main = "Ordination Centroids and Dispersion: Aitchison Distance", 
     hull=FALSE, ellipse=TRUE, label = FALSE, label.cex=0.60,
     conf = 0.68,
     pch = shape_growth.phase,
     col = color_growth.phase
     )
```


```{r}
# order factors
dispr_df <- data.frame(group = dispr$group, distances = dispr$distances)

dispr_df %>%
  ggplot(aes(x=group, y=distances, colour=group, shape=group)) +
  geom_violin(trim = FALSE) +
  geom_jitter(width=0.15) +
  scale_color_manual(values = color_growth.phase) +
  scale_shape_manual(values = shape_growth.phase) +
  scale_y_continuous(expand = c(0, 0)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

dispr_df %>%
  ggplot(aes(x=group, y=distances, colour=group, shape=group)) +
  geom_boxplot(outlier.shape=NA)+
  geom_jitter(width=0.15) +
  scale_color_manual(values = color_growth.phase) +
  scale_shape_manual(values = shape_growth.phase) +
  scale_y_continuous(expand = c(0, 0), limits = c(0, 12)) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))
```


# PERMANOVA

```{r}
permutations=10000

# data frame to store test results
# "test_id", "R2", "p_value", "p_value_dispersion")
PERMANOVA_results <- data.frame(matrix(nrow = 0, ncol = 4))
```

## Helper Functions

```{r}
permanova_on_cluster <- function(ps, test_id) {
  #Get CLR transform values
  ps_clr <- microbiome::transform(ps, "clr")
  
  #Generate distance matrix
  clr_dist_matrix <- phyloseq::distance(ps_clr, method = "euclidean") 
  
  ###############
  # ADONIS test #
  ###############
  set.seed(42)
  all_test <- vegan::adonis(clr_dist_matrix ~ phyloseq::sample_data(ps_clr)$cluster, permutations=10000)
  # print(all_test)
  
  #get dispersions
  set.seed(42)
  dispr <- vegan::betadisper(clr_dist_matrix, 
                             phyloseq::sample_data(ps_clr)$cluster,
                             bias.adjust=TRUE)
  dispr
  
  # # plot dispersions distances
  # plot(dispr, main = "Ordination Centroids and Dispersion Labeled: Aitchison Distance", sub ="")
  # 
  # # plot dispersions boxplot
  # boxplot(dispr, main = "", xlab = "")
  
  ################################################################
  # Permutation test for homogeneity of multivariate dispersions #
  ################################################################
  set.seed(42)
  pd <- permutest(dispr)
  # print(pd)
  
  R2 <- all_test[["aov.tab"]][["R2"]][1] 
  p_value <- all_test[["aov.tab"]][["Pr(>F)"]][1]
  p_value_dispersion <- pd[["tab"]][["Pr(>F)"]][1]
  res = data.frame(test_id, R2, p_value, p_value_dispersion)
  
  return(res)
}

permanova_on_growth_phase <- function(ps, test_id) {
  # Get CLR transform values
  ps_clr <- microbiome::transform(ps, "clr")
  
  #Generate distance matrix
  clr_dist_matrix <- phyloseq::distance(ps_clr, method = "euclidean") 
  
  ###############
  # ADONIS test #
  ###############
  set.seed(42)
  all_test <- vegan::adonis(clr_dist_matrix ~ phyloseq::sample_data(ps_clr)$growth.phase, permutations=10000)
  # print(all_test)
  
  #get dispersions
  set.seed(42)
  dispr <- vegan::betadisper(clr_dist_matrix, 
                             phyloseq::sample_data(ps_clr)$growth.phase,
                             bias.adjust=TRUE)
  dispr
  
  # # plot dispersions distances
  # plot(dispr, main = "Ordination Centroids and Dispersion Labeled: Aitchison Distance", sub ="")
  # 
  # # plot dispersions boxplot
  # boxplot(dispr, main = "", xlab = "")
  
  ################################################################
  # Permutation test for homogeneity of multivariate dispersions #
  ################################################################
  set.seed(42)
  pd <- permutest(dispr)
  # print(pd)
  
  R2 <- all_test[["aov.tab"]][["R2"]][1] 
  p_value <- all_test[["aov.tab"]][["Pr(>F)"]][1]
  p_value_dispersion <- pd[["tab"]][["Pr(>F)"]][1]
  res = data.frame(test_id, R2, p_value, p_value_dispersion)
  
  return(res)
}
```

## Drivers
* test at each taxrank

```{r}
rank_names(ps)
```

```{r}
drive_permanova_on_cluster <- function(ps, test_id) {
  # data frame to store test results
  results <- data.frame(matrix(nrow = 0, ncol = 4))
  
  # prune empty taxa
  ps <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
  
  # perform once at OTU level
  row <- permanova_on_cluster(ps, paste(test_id, "OTU-level", sep="_"))
  results <- rbind(results, row)
  
  # aggregate at each taxrank
  for (taxrank in rev(rank_names(ps)[-1])) {
    ps.taxa <- tax_glom(ps, taxrank = taxrank, NArm = TRUE)
    # get result of test
    row <- permanova_on_cluster(ps.taxa, paste(test_id, taxrank, sep="_"))
    results <- rbind(results, row)
  }
  
  return(results)
}


drive_permanova_on_growth_phase <- function(ps, test_id) {
  # data frame to store test results
  results <- data.frame(matrix(nrow = 0, ncol = 4))
  
  # prune empty taxa
  ps <- phyloseq::prune_taxa(phyloseq::taxa_sums(ps) > 0, ps)
  
  # perform once at OTU level
  row <- permanova_on_growth_phase(ps, paste(test_id, "OTU-level", sep="_"))
  results <- rbind(results, row)
  
  # aggregate at each taxrank
  for (taxrank in rev(rank_names(ps)[-1])) {
    ps.taxa <- tax_glom(ps, taxrank = taxrank, NArm = TRUE)
    # get result of test
    row <- permanova_on_growth_phase(ps.taxa, paste(test_id, taxrank, sep="_"))
    results <- rbind(results, row)
  }
  
  return(results)
}
```

## Cluster

```{r}
rows <- drive_permanova_on_cluster(ps, "all_cluster")
PERMANOVA_results <- rbind(PERMANOVA_results, rows)
```

## Cluster -- Pair-wise Comparisons

```{r}
set.seed(42)
combs <- combn(unique(sample_data(ps)$cluster), m=2, simplify=FALSE)
for (c in combs) {
  # get pair-wise combination
  g1 <- as.character(c[1])
  g2 <- as.character(c[2])
  test_id <- paste(g1, g2, sep=" vs ")
  # filter and run test
  print(test_id)
  rows <- drive_permanova_on_cluster(
    phyloseq::subset_samples(ps, cluster == g1 | cluster == g2),
    test_id)
  PERMANOVA_results <- rbind(PERMANOVA_results, rows)
}
```

## Focused Cluster -- Pair-wise Comparisons

```{r}
set.seed(42)
combs <- combn(unique(sample_data(ps)$growth.phase), m=2, simplify=FALSE)
for (c in combs) {
  # get pair-wise combination
  g1 <- as.character(c[1])
  g2 <- as.character(c[2])
  test_id <- paste(g1, g2, sep=" vs ")
  # filter and run test
  print(test_id)
  rows <- drive_permanova_on_growth_phase(
    phyloseq::subset_samples(ps, growth.phase == g1 | growth.phase == g2),
    test_id)
  PERMANOVA_results <- rbind(PERMANOVA_results, rows)
}
```

## Multiple Corrections

```{r}
PERMANOVA_results$adjusted_p_value <- sapply(PERMANOVA_results$p_value, p.adjust, method="BH")
```

```{r}
PERMANOVA_results$significant <- (PERMANOVA_results$adjusted_p_value < 0.05) & (PERMANOVA_results$p_value_dispersion > 0.05)
```

```{r}
PERMANOVA_results
```

### Save to .csv

```{r}
# save to csv
write.csv(PERMANOVA_results, paste(save_location, "PERMANOVA_results.csv", sep=""), quote = FALSE)
```

```{r}
# save significantly TRUE values to csv
write.csv(PERMANOVA_results[which(PERMANOVA_results$significant), ], paste(save_location, "PERMANOVA_results_TRUE.csv", sep=""), quote = FALSE)
```

