{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Data\n",
    "> By Gati Aher  \n",
    "> Sept 17, 2021\n",
    "\n",
    "**Dataset:** FCF Carbon Perturbation (Cellulose-Glucose-Malate)\n",
    "\n",
    "**Goal:**\n",
    "- Process raw data for downstream analysis\n",
    "- Put data in more workable format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import skbio\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Raw Data\n",
    "\n",
    "Inputs:\n",
    "* Sample ID: Sample Name\n",
    "    * (Jean): `sample_names.csv`\n",
    "* Sample Name x Meta Information\n",
    "    * (Jean): `FCF_annotations.csv`\n",
    "* OTU ID-Taxonomy x Sample Name x OTU Counts\n",
    "    * (MR DNA): `081616JHnew515Fcomplete-pr.fasta.otus.fa.OTU.txt`\n",
    "* OTU ID x Functions\n",
    "    * (Shree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Olin1</th>\n",
       "      <td>sC0C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin2</th>\n",
       "      <td>s1C03A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin3</th>\n",
       "      <td>s1C03B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin4</th>\n",
       "      <td>s1G03A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin5</th>\n",
       "      <td>s1G03B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin84</th>\n",
       "      <td>s3G10A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin85</th>\n",
       "      <td>s3G10B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin86</th>\n",
       "      <td>s3G10C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin87</th>\n",
       "      <td>s2M10A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin88</th>\n",
       "      <td>s2M10C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           key\n",
       "name          \n",
       "Olin1     sC0C\n",
       "Olin2   s1C03A\n",
       "Olin3   s1C03B\n",
       "Olin4   s1G03A\n",
       "Olin5   s1G03B\n",
       "...        ...\n",
       "Olin84  s3G10A\n",
       "Olin85  s3G10B\n",
       "Olin86  s3G10C\n",
       "Olin87  s2M10A\n",
       "Olin88  s2M10C\n",
       "\n",
       "[88 rows x 1 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get map of MR DNA sample IDs to sample names\n",
    "sample_names = pd.read_csv(\"../data/raw/sample_names.csv\", index_col=0)\n",
    "# prefix sample names with s (so they do not start with a number) \n",
    "sample_names['key'] = 's' + sample_names['key'].astype(str)\n",
    "\n",
    "sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>series</th>\n",
       "      <th>food</th>\n",
       "      <th>day</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C0C</th>\n",
       "      <td>C0</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>10</td>\n",
       "      <td>stable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1C03A</th>\n",
       "      <td>1C</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1C03B</th>\n",
       "      <td>1C</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1C05A</th>\n",
       "      <td>1C</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>5</td>\n",
       "      <td>transitioning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1C05B</th>\n",
       "      <td>1C</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>5</td>\n",
       "      <td>early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G07A</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>7</td>\n",
       "      <td>stabilizing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G07B</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>7</td>\n",
       "      <td>stabilizing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G07C</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>7</td>\n",
       "      <td>stabilizing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G10A</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>10</td>\n",
       "      <td>stable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G10B</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>10</td>\n",
       "      <td>stable</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      series       food  day          state\n",
       "C0C       C0  cellulose   10         stable\n",
       "1C03A     1C  cellulose    3          early\n",
       "1C03B     1C  cellulose    3          early\n",
       "1C05A     1C  cellulose    5  transitioning\n",
       "1C05B     1C  cellulose    5          early\n",
       "...      ...        ...  ...            ...\n",
       "3G07A     3G  cellulose    7    stabilizing\n",
       "3G07B     3G  cellulose    7    stabilizing\n",
       "3G07C     3G  cellulose    7    stabilizing\n",
       "3G10A     3G  cellulose   10         stable\n",
       "3G10B     3G  cellulose   10         stable\n",
       "\n",
       "[87 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get table of sample names to sample meta data\n",
    "sample_annotations_from_file = pd.read_csv(\"../data/raw/FCF_annotations.csv\", index_col=0) # these aren't in the right order\n",
    "sample_annotations_from_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>16rRNA</th>\n",
       "      <th>Percent Homology</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>homology extract gibbsfree_otu/gibbsfree_homolog</th>\n",
       "      <th>total length gibbsfree_otu/otu_lenth/gibbsfree_homolog/homolog_length</th>\n",
       "      <th>Olin1</th>\n",
       "      <th>Olin2</th>\n",
       "      <th>Olin3</th>\n",
       "      <th>Olin4</th>\n",
       "      <th>...</th>\n",
       "      <th>Olin80</th>\n",
       "      <th>Olin81</th>\n",
       "      <th>Olin82</th>\n",
       "      <th>Olin83</th>\n",
       "      <th>Olin84</th>\n",
       "      <th>Olin85</th>\n",
       "      <th>Olin86</th>\n",
       "      <th>Olin87</th>\n",
       "      <th>Olin88</th>\n",
       "      <th>Unnamed: 95</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>featureID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OTU_1</th>\n",
       "      <td>eu528231.1 seasonal dynamics mudflat mouth maj...</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>711.813612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>19</td>\n",
       "      <td>25</td>\n",
       "      <td>22</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_2</th>\n",
       "      <td>uncultured opitutus sp. ;k__bacteria;p__verruc...</td>\n",
       "      <td>96.969697</td>\n",
       "      <td>0.0</td>\n",
       "      <td>661.319285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17546</td>\n",
       "      <td>18302</td>\n",
       "      <td>32674</td>\n",
       "      <td>9700</td>\n",
       "      <td>...</td>\n",
       "      <td>31216</td>\n",
       "      <td>17838</td>\n",
       "      <td>12257</td>\n",
       "      <td>19679</td>\n",
       "      <td>16779</td>\n",
       "      <td>15614</td>\n",
       "      <td>2107</td>\n",
       "      <td>42046</td>\n",
       "      <td>56298</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_3</th>\n",
       "      <td>rhodopseudomonas palustris ;k__bacteria;p__pro...</td>\n",
       "      <td>99.493671</td>\n",
       "      <td>0.0</td>\n",
       "      <td>704.600137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3169</td>\n",
       "      <td>3033</td>\n",
       "      <td>4495</td>\n",
       "      <td>3455</td>\n",
       "      <td>...</td>\n",
       "      <td>1073</td>\n",
       "      <td>4207</td>\n",
       "      <td>1275</td>\n",
       "      <td>2274</td>\n",
       "      <td>4900</td>\n",
       "      <td>2027</td>\n",
       "      <td>1140</td>\n",
       "      <td>3192</td>\n",
       "      <td>2086</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_4</th>\n",
       "      <td>ay297802.1 waterlogged archaeological wood clo...</td>\n",
       "      <td>99.492386</td>\n",
       "      <td>0.0</td>\n",
       "      <td>702.796768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5243</td>\n",
       "      <td>3526</td>\n",
       "      <td>4658</td>\n",
       "      <td>29020</td>\n",
       "      <td>...</td>\n",
       "      <td>17928</td>\n",
       "      <td>19104</td>\n",
       "      <td>11308</td>\n",
       "      <td>11368</td>\n",
       "      <td>26958</td>\n",
       "      <td>15939</td>\n",
       "      <td>3121</td>\n",
       "      <td>30080</td>\n",
       "      <td>23869</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_5</th>\n",
       "      <td>rhodoblastus acidophilus ;k__bacteria;p__prote...</td>\n",
       "      <td>99.746835</td>\n",
       "      <td>0.0</td>\n",
       "      <td>708.206874</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3809</td>\n",
       "      <td>1128</td>\n",
       "      <td>1313</td>\n",
       "      <td>1001</td>\n",
       "      <td>...</td>\n",
       "      <td>17129</td>\n",
       "      <td>22577</td>\n",
       "      <td>7156</td>\n",
       "      <td>26149</td>\n",
       "      <td>9243</td>\n",
       "      <td>2211</td>\n",
       "      <td>1368</td>\n",
       "      <td>13865</td>\n",
       "      <td>5190</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 95 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      16rRNA  \\\n",
       "featureID                                                      \n",
       "OTU_1      eu528231.1 seasonal dynamics mudflat mouth maj...   \n",
       "OTU_2      uncultured opitutus sp. ;k__bacteria;p__verruc...   \n",
       "OTU_3      rhodopseudomonas palustris ;k__bacteria;p__pro...   \n",
       "OTU_4      ay297802.1 waterlogged archaeological wood clo...   \n",
       "OTU_5      rhodoblastus acidophilus ;k__bacteria;p__prote...   \n",
       "\n",
       "           Percent Homology  evalue    bitscore  \\\n",
       "featureID                                         \n",
       "OTU_1            100.000000     0.0  711.813612   \n",
       "OTU_2             96.969697     0.0  661.319285   \n",
       "OTU_3             99.493671     0.0  704.600137   \n",
       "OTU_4             99.492386     0.0  702.796768   \n",
       "OTU_5             99.746835     0.0  708.206874   \n",
       "\n",
       "           homology extract gibbsfree_otu/gibbsfree_homolog  \\\n",
       "featureID                                                     \n",
       "OTU_1                                                   NaN   \n",
       "OTU_2                                                   NaN   \n",
       "OTU_3                                                   NaN   \n",
       "OTU_4                                                   NaN   \n",
       "OTU_5                                                   NaN   \n",
       "\n",
       "           total length gibbsfree_otu/otu_lenth/gibbsfree_homolog/homolog_length  \\\n",
       "featureID                                                                          \n",
       "OTU_1                                                    NaN                       \n",
       "OTU_2                                                    NaN                       \n",
       "OTU_3                                                    NaN                       \n",
       "OTU_4                                                    NaN                       \n",
       "OTU_5                                                    NaN                       \n",
       "\n",
       "           Olin1  Olin2  Olin3  Olin4  ...  Olin80  Olin81  Olin82  Olin83  \\\n",
       "featureID                              ...                                   \n",
       "OTU_1          6     11     13     20  ...      11      19      25      22   \n",
       "OTU_2      17546  18302  32674   9700  ...   31216   17838   12257   19679   \n",
       "OTU_3       3169   3033   4495   3455  ...    1073    4207    1275    2274   \n",
       "OTU_4       5243   3526   4658  29020  ...   17928   19104   11308   11368   \n",
       "OTU_5       3809   1128   1313   1001  ...   17129   22577    7156   26149   \n",
       "\n",
       "           Olin84  Olin85  Olin86  Olin87  Olin88  Unnamed: 95  \n",
       "featureID                                                       \n",
       "OTU_1          16      16       6       5       4          NaN  \n",
       "OTU_2       16779   15614    2107   42046   56298          NaN  \n",
       "OTU_3        4900    2027    1140    3192    2086          NaN  \n",
       "OTU_4       26958   15939    3121   30080   23869          NaN  \n",
       "OTU_5        9243    2211    1368   13865    5190          NaN  \n",
       "\n",
       "[5 rows x 95 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get table of OTU ID-taxonomy x absolute OTU counts\n",
    "otu_raw = pd.read_csv(\"../data/raw/081616JHnew515Fcomplete-pr.fasta.otus.fa.OTU.txt\", sep=\"\\t\", index_col=0)\n",
    "otu_raw.index.names = ['featureID']\n",
    "otu_raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get table of OTU functions\n",
    "# TODO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create and Save Processed Data\n",
    "\n",
    "Outputs:\n",
    "* Sample Name x Meta Information\n",
    "    * `sample_metadata.tsv`\n",
    "* OTU ID x Sample ID x HTS Counts\n",
    "    * `OTU_counts.tsv`\n",
    "* OTU ID x Taxonomy, Functions\n",
    "    * `feature_metadata.tsv`\n",
    "* OTU ID x Taxonomy\n",
    "    * `taxonomy.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create `sample_metadata.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samples: ['C0C', '1C', '1C', '1C', '1C', '1C', '1C', '1C', '1C', '1C', '1C', '2C', '2C', '2C', '2C', '2C', '2C', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '3G', '3G', '3G', '3G', '3G', '3G', '3G', '3G', '3G', '3G', '3G']\n"
     ]
    }
   ],
   "source": [
    "# format useful sample annotations file\n",
    "# columns: index series carbon transfer day\n",
    "\n",
    "ann_index = []\n",
    "ann_series = []\n",
    "ann_carbon = []\n",
    "ann_transfer = []\n",
    "ann_group = []\n",
    "ann_day = []\n",
    "ann_replicate = []\n",
    "ann_state = []\n",
    "\n",
    "for sn in sample_annotations_from_file.index:\n",
    "    # index\n",
    "    ann_index.append(\"s\" + sn)\n",
    "    # series\n",
    "    if sn == \"C0C\":\n",
    "        ann_series.append(\"C0C\")\n",
    "    else:\n",
    "        ann_series.append(\"s\" + sn[1])\n",
    "    # carbon\n",
    "    if sn == \"C0C\":\n",
    "        ann_carbon.append(\"original\")\n",
    "    elif sample_annotations_from_file.loc[sn, \"food\"] == \"cellulose\":\n",
    "        ann_carbon.append(\"cellulose\")\n",
    "    elif sample_annotations_from_file.loc[sn, \"food\"] == \"glucose\":\n",
    "        ann_carbon.append(\"glucose\")\n",
    "    elif sample_annotations_from_file.loc[sn, \"food\"] == \"malate\":\n",
    "        ann_carbon.append(\"malate\")\n",
    "    else:\n",
    "        ann_carbon.append(\"X\")\n",
    "    # transfer\n",
    "    if sn == \"C0C\":\n",
    "        ann_transfer.append(0)\n",
    "    else:\n",
    "        ann_transfer.append(int(sn[0]))\n",
    "    # group\n",
    "    if sn == \"C0C\":\n",
    "        ann_group.append(\"C0C\")\n",
    "    else:\n",
    "        ann_group.append(sn[0:2])\n",
    "    # day\n",
    "    if sn == \"C0C\":\n",
    "        ann_day.append(10)\n",
    "    else:\n",
    "        ann_day.append(int(sn[2:-1]))\n",
    "    # replicate\n",
    "    if sn == \"C0C\":\n",
    "        ann_replicate.append(\"C0C\")\n",
    "    else:\n",
    "        ann_replicate.append(sn[-1])  \n",
    "    # state (from unsupervised clustering)\n",
    "    ann_state.append(sample_annotations_from_file.loc[sn, \"state\"])\n",
    "        \n",
    "# print(\"ann_group\", ann_group)\n",
    "# print(\"ann_series\", ann_series)\n",
    "# print(\"ann_carbon\", ann_carbon)\n",
    "# print(\"ann_transfer\", ann_transfer)\n",
    "# print(\"ann_day\", ann_day)\n",
    "# print(\"ann_replicate\", ann_replicate)\n",
    "# print(\"ann_index\", ann_index)\n",
    "\n",
    "sample_metadata = pd.DataFrame()\n",
    "sample_metadata[\"group\"] = ann_group\n",
    "sample_metadata[\"series\"] = ann_series\n",
    "sample_metadata[\"carbon\"] = ann_carbon\n",
    "sample_metadata[\"transfer\"] = ann_transfer\n",
    "sample_metadata[\"day\"] = ann_day\n",
    "sample_metadata[\"replicate\"] = ann_replicate\n",
    "sample_metadata[\"state\"] = ann_state\n",
    "sample_metadata[\"sampleID\"] = ann_index\n",
    "sample_metadata = sample_metadata.set_index(\"sampleID\")\n",
    "\n",
    "print(\"samples:\", list(sample_metadata.group))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>series</th>\n",
       "      <th>carbon</th>\n",
       "      <th>transfer</th>\n",
       "      <th>day</th>\n",
       "      <th>replicate</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sC0C</th>\n",
       "      <td>C0C</td>\n",
       "      <td>C0C</td>\n",
       "      <td>original</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>C0C</td>\n",
       "      <td>stable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s1C03A</th>\n",
       "      <td>1C</td>\n",
       "      <td>sC</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s1C03B</th>\n",
       "      <td>1C</td>\n",
       "      <td>sC</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s1C05A</th>\n",
       "      <td>1C</td>\n",
       "      <td>sC</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>transitioning</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s1C05B</th>\n",
       "      <td>1C</td>\n",
       "      <td>sC</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>B</td>\n",
       "      <td>early</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G07A</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>A</td>\n",
       "      <td>stabilizing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G07B</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>B</td>\n",
       "      <td>stabilizing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G07C</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>C</td>\n",
       "      <td>stabilizing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G10A</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>A</td>\n",
       "      <td>stable</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G10B</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>B</td>\n",
       "      <td>stable</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         group series     carbon  transfer  day replicate          state\n",
       "sampleID                                                                \n",
       "sC0C       C0C    C0C   original         0   10       C0C         stable\n",
       "s1C03A      1C     sC  cellulose         1    3         A          early\n",
       "s1C03B      1C     sC  cellulose         1    3         B          early\n",
       "s1C05A      1C     sC  cellulose         1    5         A  transitioning\n",
       "s1C05B      1C     sC  cellulose         1    5         B          early\n",
       "...        ...    ...        ...       ...  ...       ...            ...\n",
       "s3G07A      3G     sG  cellulose         3    7         A    stabilizing\n",
       "s3G07B      3G     sG  cellulose         3    7         B    stabilizing\n",
       "s3G07C      3G     sG  cellulose         3    7         C    stabilizing\n",
       "s3G10A      3G     sG  cellulose         3   10         A         stable\n",
       "s3G10B      3G     sG  cellulose         3   10         B         stable\n",
       "\n",
       "[86 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_metadata = sample_metadata.drop([\"s2M07C\"])\n",
    "sample_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "numpy.int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(sample_metadata[\"day\"][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "sample_metadata.to_csv(\"../data/processed/sample_metadata.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create `OTU_counts.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# format table of OTU ID x sample name x absolute counts\n",
    "OTU_counts = otu_raw.filter(regex='^Olin')\n",
    "# rename sample ID to sample names\n",
    "OTU_counts = OTU_counts.rename(columns=sample_names.to_dict()[\"key\"])\n",
    "# NOTE: 3G10C and 2M7C were removed because they were incorrectly sequenced\n",
    "OTU_counts = OTU_counts.drop(columns=[\"s3G10C\", \"s2M07C\"])\n",
    "# order columns by order in sample metadata\n",
    "OTU_counts = OTU_counts.reindex(columns=list(sample_metadata.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "OTU_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "OTU_counts.to_csv(\"../data/processed/OTU_counts.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create `feature_metadata.tsv` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# format table of OTU ID x taxonomy & function\n",
    "feature_metadata = pd.DataFrame(index=otu_raw.index, columns=[\"Name\"])\n",
    "\n",
    "for otu, desc in otu_raw.iterrows(): \n",
    "    parts = desc[\"16rRNA\"].split(\";\")\n",
    "    taxonomy = \"Unassigned\"\n",
    "    for i, part in enumerate(parts):\n",
    "        if i == 0:\n",
    "            feature_metadata.loc[otu, \"Name\"] = part\n",
    "        elif part.startswith(\"k__\"):\n",
    "            feature_metadata.loc[otu, \"kingdom\"] = part[3:]\n",
    "            taxonomy = \"Root;\" + part + \";\"\n",
    "        elif part.startswith(\"p__\"):\n",
    "            feature_metadata.loc[otu, \"phylum\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"c__\"):\n",
    "            feature_metadata.loc[otu, \"class\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"o__\"):\n",
    "            feature_metadata.loc[otu, \"order\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"f__\"):\n",
    "            feature_metadata.loc[otu, \"family\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"g__\"):\n",
    "            feature_metadata.loc[otu, \"genus\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"s__\"):\n",
    "            feature_metadata.loc[otu, \"species\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"superkingdom__\"):\n",
    "            feature_metadata.loc[otu, \"superkingdom\"] = part[14:]\n",
    "        elif part.startswith(\"superphylum__\"):\n",
    "            feature_metadata.loc[otu, \"superphylum\"] = part[13:]\n",
    "        elif part.startswith(\"subclass\"):\n",
    "            feature_metadata.loc[otu, \"subclass\"] = part[8:]\n",
    "        elif part.startswith(\"suborder\"):\n",
    "            feature_metadata.loc[otu, \"suborder\"] = part[8:]\n",
    "        elif part.startswith(\"subphylum__\"):\n",
    "            feature_metadata.loc[otu, \"subphylum\"] = part[9:]\n",
    "        else:\n",
    "            print(\"OTU:\", otu, \"Extra PART:\", part)\n",
    "    feature_metadata.loc[otu, \"taxonomy\"] = taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "feature_metadata.to_csv(\"../data/processed/feature_metadata.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create `taxonomy.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "feature_metadata[\"taxonomy\"].to_csv(\"../data/processed/taxonomy.tsv\", sep='\\t')\n",
    "feature_metadata[\"taxonomy\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Additional Cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check if there are OTUs that appear in no samples\n",
    "OTU_counts[OTU_counts.sum(axis=1) == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove OTUs that are not bacteria\n",
    "not_bacteria = feature_metadata[feature_metadata[\"kingdom\"] != \"bacteria\"]\n",
    "print(not_bacteria.Name)\n",
    "\n",
    "OTU_counts_clean = OTU_counts.drop(not_bacteria.index)\n",
    "print(\"OTU_counts.shape\", OTU_counts.shape, \"OTU_counts_clean.shape\", OTU_counts_clean.shape,)\n",
    "\n",
    "feature_metadata_clean = feature_metadata.drop(not_bacteria.index)\n",
    "print(\"feature_metadata.shape\", feature_metadata.shape, \"feature_metadata_clean.shape\", feature_metadata_clean.shape,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "not_bacteria.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove OTUs with less than threshold total reads\n",
    "threshold_OTU_total_reads = 5\n",
    "less_than_threshold = OTU_counts_clean[OTU_counts_clean.sum(axis=1) < threshold_OTU_total_reads]\n",
    "print(less_than_threshold.sum(axis=1))\n",
    "\n",
    "OTU_counts_clean = OTU_counts_clean.drop(less_than_threshold.index)\n",
    "print(\"OTU_counts_clean.shape\", OTU_counts_clean.shape,)\n",
    "\n",
    "feature_metadata_clean = feature_metadata_clean.drop(less_than_threshold.index)\n",
    "print(\"feature_metadata_clean.shape\", feature_metadata_clean.shape,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save clean\n",
    "OTU_counts_clean.to_csv(\"../data/processed/OTU_counts_clean.tsv\", sep='\\t')\n",
    "feature_metadata_clean.to_csv(\"../data/processed/feature_metadata_clean.tsv\", sep='\\t')\n",
    "feature_metadata_clean[[\"kingdom\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"]].to_csv(\"../data/processed/taxonomy_table_clean.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get Phylogenetic Info and Alignment Scores\n",
    "\n",
    "* OTU ID: DNA Sequences\n",
    "    * (MR DNA): `081616JHnew515Fcomplete-pr.fasta.otus.fa`\n",
    "* OTU ID x OTU ID Pairwise Alignment Scores\n",
    "    * (Muscle): `Pairwise_distances_all_OTU_Muscle`\n",
    "* OTU ID Phylogenetic Tree\n",
    "    * (Muscle): `MUSCLE_alignment_ML_tree_all_OTUs`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rename OTU ID: DNA Sequences mapping (for Jean, to visually analyze phylogenetic tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from skbio.sequence import DNA\n",
    "\n",
    "sequence_file = open(\"../data/raw/081616JHnew515Fcomplete-pr.fasta.otus.fa\", 'r')\n",
    "sequence_lines = [l.upper() for l in sequence_file.readlines()]\n",
    "dna_lines = [DNA(str(seq), metadata=seq.metadata) for seq in skbio.io.read(sequence_lines, format='fasta')]\n",
    "dna_lines_renamed = []\n",
    "\n",
    "for seq in skbio.io.read(sequence_lines, format='fasta'):\n",
    "    str_seq = str(seq)\n",
    "    metadata = seq.metadata\n",
    "    metadata['id'] = metadata['id'] + \"_\" + feature_metadata.loc[metadata['id'], \"Name\"].replace(\" \", \"_\").replace(\".\", \"\")\n",
    "    dna_lines_renamed.append(DNA(str_seq, metadata=metadata))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "dna_lines_renamed_file = open(\"../data/processed/renamed_081616JHnew515Fcomplete-pr.fasta.otus.fa\", \"w\")\n",
    "for seq in dna_lines_renamed:\n",
    "    skbio.io.write(seq, into=dna_lines_renamed_file, format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import and Use Newick Phylogenetic Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import phylogenetic weights and use in an example\n",
    "from skbio import TreeNode\n",
    "tree_file = open(\"../data/raw/MUSCLE_alignment_ML_tree_all_OTUs\", \"r\")\n",
    "tree = TreeNode.read(tree_file)\n",
    "tree_rooted = tree.root_at_midpoint()\n",
    "print(str(tree)[:50], \"...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# alpha diversity w/ phylogenetic weights example \n",
    "from skbio.diversity import alpha_diversity\n",
    "alpha_diversity(metric=\"faith_pd\", counts=OTU_counts.T.values, ids=OTU_counts.columns, tree=tree_rooted, otu_ids=[n.name for n in tree_rooted.tips()], validate=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experiment Scratch-Pad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
