{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Data\n",
    "> By Gati Aher  \n",
    "> Sept 17, 2021\n",
    "\n",
    "**Dataset:** FCF Carbon Perturbation (Cellulose-Glucose-Malate)\n",
    "\n",
    "**Goal:**\n",
    "- Process raw data for downstream analysis\n",
    "- Put data in more workable format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# imports\n",
    "import skbio\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load Raw Data\n",
    "\n",
    "Inputs:\n",
    "* Sample ID: Sample Name\n",
    "    * (Jean): `sample_names.csv`\n",
    "* Sample Name x Meta Information\n",
    "    * (Jean): `FCF_annotations.csv`\n",
    "* OTU ID-Taxonomy x Sample Name x OTU Counts\n",
    "    * (MR DNA): `081616JHnew515Fcomplete-pr.fasta.otus.fa.OTU.txt`\n",
    "* OTU ID x Functions\n",
    "    * (Shree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Olin1</th>\n",
       "      <td>sC0C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin2</th>\n",
       "      <td>s1C3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin3</th>\n",
       "      <td>s1C3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin4</th>\n",
       "      <td>s1G3A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin5</th>\n",
       "      <td>s1G3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin84</th>\n",
       "      <td>s3G10A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin85</th>\n",
       "      <td>s3G10B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin86</th>\n",
       "      <td>s3G10C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin87</th>\n",
       "      <td>s2M10A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Olin88</th>\n",
       "      <td>s2M10C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>88 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           key\n",
       "name          \n",
       "Olin1     sC0C\n",
       "Olin2    s1C3A\n",
       "Olin3    s1C3B\n",
       "Olin4    s1G3A\n",
       "Olin5    s1G3B\n",
       "...        ...\n",
       "Olin84  s3G10A\n",
       "Olin85  s3G10B\n",
       "Olin86  s3G10C\n",
       "Olin87  s2M10A\n",
       "Olin88  s2M10C\n",
       "\n",
       "[88 rows x 1 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get map of MR DNA sample IDs to sample names\n",
    "sample_names = pd.read_csv(\"../data/raw/sample_names.csv\", index_col=0)\n",
    "# prefix sample names with s (so they do not start with a number) \n",
    "sample_names['key'] = 's' + sample_names['key'].astype(str)\n",
    "\n",
    "sample_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>series</th>\n",
       "      <th>food</th>\n",
       "      <th>day</th>\n",
       "      <th>gradient</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C0C</th>\n",
       "      <td>C0</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1C3A</th>\n",
       "      <td>1C</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1C3B</th>\n",
       "      <td>1C</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1C5A</th>\n",
       "      <td>1C</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1C5B</th>\n",
       "      <td>1C</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G7A</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G7B</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G7C</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G10A</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3G10B</th>\n",
       "      <td>3G</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      series       food  day  gradient\n",
       "C0C       C0  cellulose   10         0\n",
       "1C3A      1C  cellulose    3         5\n",
       "1C3B      1C  cellulose    3         5\n",
       "1C5A      1C  cellulose    5         4\n",
       "1C5B      1C  cellulose    5         4\n",
       "...      ...        ...  ...       ...\n",
       "3G7A      3G  cellulose    7         3\n",
       "3G7B      3G  cellulose    7         3\n",
       "3G7C      3G  cellulose    7         3\n",
       "3G10A     3G  cellulose   10         2\n",
       "3G10B     3G  cellulose   10         2\n",
       "\n",
       "[87 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get table of sample names to sample meta data\n",
    "sample_annotations_from_file = pd.read_csv(\"../data/raw/FCF_annotations.csv\", index_col=0) # these aren't in the right order\n",
    "sample_annotations_from_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>16rRNA</th>\n",
       "      <th>Percent Homology</th>\n",
       "      <th>evalue</th>\n",
       "      <th>bitscore</th>\n",
       "      <th>homology extract gibbsfree_otu/gibbsfree_homolog</th>\n",
       "      <th>total length gibbsfree_otu/otu_lenth/gibbsfree_homolog/homolog_length</th>\n",
       "      <th>Olin1</th>\n",
       "      <th>Olin2</th>\n",
       "      <th>Olin3</th>\n",
       "      <th>Olin4</th>\n",
       "      <th>...</th>\n",
       "      <th>Olin80</th>\n",
       "      <th>Olin81</th>\n",
       "      <th>Olin82</th>\n",
       "      <th>Olin83</th>\n",
       "      <th>Olin84</th>\n",
       "      <th>Olin85</th>\n",
       "      <th>Olin86</th>\n",
       "      <th>Olin87</th>\n",
       "      <th>Olin88</th>\n",
       "      <th>Unnamed: 95</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>featureID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OTU_1</th>\n",
       "      <td>eu528231.1 seasonal dynamics mudflat mouth maj...</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>711.813612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>20</td>\n",
       "      <td>...</td>\n",
       "      <td>11</td>\n",
       "      <td>19</td>\n",
       "      <td>25</td>\n",
       "      <td>22</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_2</th>\n",
       "      <td>uncultured opitutus sp. ;k__bacteria;p__verruc...</td>\n",
       "      <td>96.969697</td>\n",
       "      <td>0.0</td>\n",
       "      <td>661.319285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17546</td>\n",
       "      <td>18302</td>\n",
       "      <td>32674</td>\n",
       "      <td>9700</td>\n",
       "      <td>...</td>\n",
       "      <td>31216</td>\n",
       "      <td>17838</td>\n",
       "      <td>12257</td>\n",
       "      <td>19679</td>\n",
       "      <td>16779</td>\n",
       "      <td>15614</td>\n",
       "      <td>2107</td>\n",
       "      <td>42046</td>\n",
       "      <td>56298</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_3</th>\n",
       "      <td>rhodopseudomonas palustris ;k__bacteria;p__pro...</td>\n",
       "      <td>99.493671</td>\n",
       "      <td>0.0</td>\n",
       "      <td>704.600137</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3169</td>\n",
       "      <td>3033</td>\n",
       "      <td>4495</td>\n",
       "      <td>3455</td>\n",
       "      <td>...</td>\n",
       "      <td>1073</td>\n",
       "      <td>4207</td>\n",
       "      <td>1275</td>\n",
       "      <td>2274</td>\n",
       "      <td>4900</td>\n",
       "      <td>2027</td>\n",
       "      <td>1140</td>\n",
       "      <td>3192</td>\n",
       "      <td>2086</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_4</th>\n",
       "      <td>ay297802.1 waterlogged archaeological wood clo...</td>\n",
       "      <td>99.492386</td>\n",
       "      <td>0.0</td>\n",
       "      <td>702.796768</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5243</td>\n",
       "      <td>3526</td>\n",
       "      <td>4658</td>\n",
       "      <td>29020</td>\n",
       "      <td>...</td>\n",
       "      <td>17928</td>\n",
       "      <td>19104</td>\n",
       "      <td>11308</td>\n",
       "      <td>11368</td>\n",
       "      <td>26958</td>\n",
       "      <td>15939</td>\n",
       "      <td>3121</td>\n",
       "      <td>30080</td>\n",
       "      <td>23869</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_5</th>\n",
       "      <td>rhodoblastus acidophilus ;k__bacteria;p__prote...</td>\n",
       "      <td>99.746835</td>\n",
       "      <td>0.0</td>\n",
       "      <td>708.206874</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3809</td>\n",
       "      <td>1128</td>\n",
       "      <td>1313</td>\n",
       "      <td>1001</td>\n",
       "      <td>...</td>\n",
       "      <td>17129</td>\n",
       "      <td>22577</td>\n",
       "      <td>7156</td>\n",
       "      <td>26149</td>\n",
       "      <td>9243</td>\n",
       "      <td>2211</td>\n",
       "      <td>1368</td>\n",
       "      <td>13865</td>\n",
       "      <td>5190</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 95 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                      16rRNA  \\\n",
       "featureID                                                      \n",
       "OTU_1      eu528231.1 seasonal dynamics mudflat mouth maj...   \n",
       "OTU_2      uncultured opitutus sp. ;k__bacteria;p__verruc...   \n",
       "OTU_3      rhodopseudomonas palustris ;k__bacteria;p__pro...   \n",
       "OTU_4      ay297802.1 waterlogged archaeological wood clo...   \n",
       "OTU_5      rhodoblastus acidophilus ;k__bacteria;p__prote...   \n",
       "\n",
       "           Percent Homology  evalue    bitscore  \\\n",
       "featureID                                         \n",
       "OTU_1            100.000000     0.0  711.813612   \n",
       "OTU_2             96.969697     0.0  661.319285   \n",
       "OTU_3             99.493671     0.0  704.600137   \n",
       "OTU_4             99.492386     0.0  702.796768   \n",
       "OTU_5             99.746835     0.0  708.206874   \n",
       "\n",
       "           homology extract gibbsfree_otu/gibbsfree_homolog  \\\n",
       "featureID                                                     \n",
       "OTU_1                                                   NaN   \n",
       "OTU_2                                                   NaN   \n",
       "OTU_3                                                   NaN   \n",
       "OTU_4                                                   NaN   \n",
       "OTU_5                                                   NaN   \n",
       "\n",
       "           total length gibbsfree_otu/otu_lenth/gibbsfree_homolog/homolog_length  \\\n",
       "featureID                                                                          \n",
       "OTU_1                                                    NaN                       \n",
       "OTU_2                                                    NaN                       \n",
       "OTU_3                                                    NaN                       \n",
       "OTU_4                                                    NaN                       \n",
       "OTU_5                                                    NaN                       \n",
       "\n",
       "           Olin1  Olin2  Olin3  Olin4  ...  Olin80  Olin81  Olin82  Olin83  \\\n",
       "featureID                              ...                                   \n",
       "OTU_1          6     11     13     20  ...      11      19      25      22   \n",
       "OTU_2      17546  18302  32674   9700  ...   31216   17838   12257   19679   \n",
       "OTU_3       3169   3033   4495   3455  ...    1073    4207    1275    2274   \n",
       "OTU_4       5243   3526   4658  29020  ...   17928   19104   11308   11368   \n",
       "OTU_5       3809   1128   1313   1001  ...   17129   22577    7156   26149   \n",
       "\n",
       "           Olin84  Olin85  Olin86  Olin87  Olin88  Unnamed: 95  \n",
       "featureID                                                       \n",
       "OTU_1          16      16       6       5       4          NaN  \n",
       "OTU_2       16779   15614    2107   42046   56298          NaN  \n",
       "OTU_3        4900    2027    1140    3192    2086          NaN  \n",
       "OTU_4       26958   15939    3121   30080   23869          NaN  \n",
       "OTU_5        9243    2211    1368   13865    5190          NaN  \n",
       "\n",
       "[5 rows x 95 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get table of OTU ID-taxonomy x absolute OTU counts\n",
    "otu_raw = pd.read_csv(\"../data/raw/081616JHnew515Fcomplete-pr.fasta.otus.fa.OTU.txt\", sep=\"\\t\", index_col=0)\n",
    "otu_raw.index.names = ['featureID']\n",
    "otu_raw.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get table of OTU functions\n",
    "# TODO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create and Save Processed Data\n",
    "\n",
    "Outputs:\n",
    "* Sample Name x Meta Information\n",
    "    * `sample_metadata.tsv`\n",
    "* OTU ID x Sample ID x HTS Counts\n",
    "    * `OTU_counts.tsv`\n",
    "* OTU ID x Taxonomy, Functions\n",
    "    * `feature_metadata.tsv`\n",
    "* OTU ID x Taxonomy\n",
    "    * `taxonomy.tsv`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create `sample_metadata.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samples: ['C0C', '1C', '1C', '1C', '1C', '1C', '1C', '1C', '1C', '1C', '1C', '2C', '2C', '2C', '2C', '2C', '2C', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '1M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '2M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '3M', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '1G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '2G', '3G', '3G', '3G', '3G', '3G', '3G', '3G', '3G', '3G', '3G', '3G']\n"
     ]
    }
   ],
   "source": [
    "# format useful sample annotations file\n",
    "# columns: index series carbon transfer day\n",
    "\n",
    "ann_index = []\n",
    "ann_series = []\n",
    "ann_carbon = []\n",
    "ann_transfer = []\n",
    "ann_group = []\n",
    "ann_day = []\n",
    "ann_replicate = []\n",
    "ann_gradient = []\n",
    "\n",
    "for sn in sample_annotations_from_file.index:\n",
    "    # index\n",
    "    ann_index.append(\"s\" + sn)\n",
    "    # series\n",
    "    if sn == \"C0C\":\n",
    "        ann_series.append(\"C0C\")\n",
    "    else:\n",
    "        ann_series.append(\"s\" + sn[1])\n",
    "    # carbon\n",
    "    if sn == \"C0C\":\n",
    "        ann_carbon.append(\"original\")\n",
    "    elif sample_annotations_from_file.loc[sn, \"food\"] == \"cellulose\":\n",
    "        ann_carbon.append(\"cellulose\")\n",
    "    elif sample_annotations_from_file.loc[sn, \"food\"] == \"glucose\":\n",
    "        ann_carbon.append(\"glucose\")\n",
    "    elif sample_annotations_from_file.loc[sn, \"food\"] == \"malate\":\n",
    "        ann_carbon.append(\"malate\")\n",
    "    else:\n",
    "        ann_carbon.append(\"X\")\n",
    "    # transfer\n",
    "    if sn == \"C0C\":\n",
    "        ann_transfer.append(0)\n",
    "    else:\n",
    "        ann_transfer.append(int(sn[0]))\n",
    "    # group\n",
    "    if sn == \"C0C\":\n",
    "        ann_group.append(\"C0C\")\n",
    "    else:\n",
    "        ann_group.append(sn[0:2])\n",
    "    # day\n",
    "    if sn == \"C0C\":\n",
    "        ann_day.append(10)\n",
    "    else:\n",
    "        ann_day.append(int(sn[2:-1]))\n",
    "    # replicate\n",
    "    if sn == \"C0C\":\n",
    "        ann_replicate.append(\"C0C\")\n",
    "    else:\n",
    "        ann_replicate.append(sn[-1])  \n",
    "    # gradient\n",
    "    ann_gradient.append(sample_annotations_from_file.loc[sn, \"gradient\"])\n",
    "        \n",
    "# print(\"ann_group\", ann_group)\n",
    "# print(\"ann_series\", ann_series)\n",
    "# print(\"ann_carbon\", ann_carbon)\n",
    "# print(\"ann_transfer\", ann_transfer)\n",
    "# print(\"ann_day\", ann_day)\n",
    "# print(\"ann_replicate\", ann_replicate)\n",
    "# print(\"ann_index\", ann_index)\n",
    "\n",
    "sample_metadata = pd.DataFrame()\n",
    "sample_metadata[\"group\"] = ann_group\n",
    "sample_metadata[\"series\"] = ann_series\n",
    "sample_metadata[\"carbon\"] = ann_carbon\n",
    "sample_metadata[\"transfer\"] = ann_transfer\n",
    "sample_metadata[\"day\"] = ann_day\n",
    "sample_metadata[\"replicate\"] = ann_replicate\n",
    "sample_metadata[\"gradient\"] = ann_gradient\n",
    "sample_metadata[\"sampleID\"] = ann_index\n",
    "sample_metadata = sample_metadata.set_index(\"sampleID\")\n",
    "\n",
    "print(\"samples:\", list(sample_metadata.group))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>group</th>\n",
       "      <th>series</th>\n",
       "      <th>carbon</th>\n",
       "      <th>transfer</th>\n",
       "      <th>day</th>\n",
       "      <th>replicate</th>\n",
       "      <th>gradient</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sampleID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sC0C</th>\n",
       "      <td>C0C</td>\n",
       "      <td>C0C</td>\n",
       "      <td>original</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>C0C</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s1C3A</th>\n",
       "      <td>1C</td>\n",
       "      <td>sC</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>A</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s1C3B</th>\n",
       "      <td>1C</td>\n",
       "      <td>sC</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>B</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s1C5A</th>\n",
       "      <td>1C</td>\n",
       "      <td>sC</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>A</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s1C5B</th>\n",
       "      <td>1C</td>\n",
       "      <td>sC</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>B</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G7A</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>A</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G7B</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>B</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G7C</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>C</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G10A</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>s3G10B</th>\n",
       "      <td>3G</td>\n",
       "      <td>sG</td>\n",
       "      <td>cellulose</td>\n",
       "      <td>3</td>\n",
       "      <td>10</td>\n",
       "      <td>B</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>86 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         group series     carbon  transfer  day replicate  gradient\n",
       "sampleID                                                           \n",
       "sC0C       C0C    C0C   original         0   10       C0C         0\n",
       "s1C3A       1C     sC  cellulose         1    3         A         5\n",
       "s1C3B       1C     sC  cellulose         1    3         B         5\n",
       "s1C5A       1C     sC  cellulose         1    5         A         4\n",
       "s1C5B       1C     sC  cellulose         1    5         B         4\n",
       "...        ...    ...        ...       ...  ...       ...       ...\n",
       "s3G7A       3G     sG  cellulose         3    7         A         3\n",
       "s3G7B       3G     sG  cellulose         3    7         B         3\n",
       "s3G7C       3G     sG  cellulose         3    7         C         3\n",
       "s3G10A      3G     sG  cellulose         3   10         A         2\n",
       "s3G10B      3G     sG  cellulose         3   10         B         2\n",
       "\n",
       "[86 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_metadata = sample_metadata.drop([\"s2M7C\"])\n",
    "sample_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "numpy.int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(sample_metadata[\"day\"][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "sample_metadata.to_csv(\"../data/processed/sample_metadata.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create `OTU_counts.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# format table of OTU ID x sample name x absolute counts\n",
    "OTU_counts = otu_raw.filter(regex='^Olin')\n",
    "# rename sample ID to sample names\n",
    "OTU_counts = OTU_counts.rename(columns=sample_names.to_dict()[\"key\"])\n",
    "# NOTE: 3G10C and 2M7C were removed because they were incorrectly sequenced\n",
    "OTU_counts = OTU_counts.drop(columns=[\"s3G10C\", \"s2M7C\"])\n",
    "# order columns by order in sample metadata\n",
    "OTU_counts = OTU_counts.reindex(columns=list(sample_metadata.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sC0C</th>\n",
       "      <th>s1C3A</th>\n",
       "      <th>s1C3B</th>\n",
       "      <th>s1C5A</th>\n",
       "      <th>s1C5B</th>\n",
       "      <th>s1C7A</th>\n",
       "      <th>s1C7B</th>\n",
       "      <th>s1C10A</th>\n",
       "      <th>s1C10B</th>\n",
       "      <th>s1C20A</th>\n",
       "      <th>...</th>\n",
       "      <th>s3G3B</th>\n",
       "      <th>s3G3C</th>\n",
       "      <th>s3G5A</th>\n",
       "      <th>s3G5B</th>\n",
       "      <th>s3G5C</th>\n",
       "      <th>s3G7A</th>\n",
       "      <th>s3G7B</th>\n",
       "      <th>s3G7C</th>\n",
       "      <th>s3G10A</th>\n",
       "      <th>s3G10B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>featureID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OTU_1</th>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>30</td>\n",
       "      <td>15</td>\n",
       "      <td>23</td>\n",
       "      <td>25</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>30</td>\n",
       "      <td>46</td>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>10</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_2</th>\n",
       "      <td>17546</td>\n",
       "      <td>18302</td>\n",
       "      <td>32674</td>\n",
       "      <td>15446</td>\n",
       "      <td>22216</td>\n",
       "      <td>25011</td>\n",
       "      <td>21224</td>\n",
       "      <td>13067</td>\n",
       "      <td>13340</td>\n",
       "      <td>6896</td>\n",
       "      <td>...</td>\n",
       "      <td>6869</td>\n",
       "      <td>8322</td>\n",
       "      <td>9792</td>\n",
       "      <td>21106</td>\n",
       "      <td>39348</td>\n",
       "      <td>24470</td>\n",
       "      <td>34872</td>\n",
       "      <td>33561</td>\n",
       "      <td>16779</td>\n",
       "      <td>15614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_3</th>\n",
       "      <td>3169</td>\n",
       "      <td>3033</td>\n",
       "      <td>4495</td>\n",
       "      <td>3138</td>\n",
       "      <td>4601</td>\n",
       "      <td>3633</td>\n",
       "      <td>1919</td>\n",
       "      <td>2952</td>\n",
       "      <td>2110</td>\n",
       "      <td>5739</td>\n",
       "      <td>...</td>\n",
       "      <td>610</td>\n",
       "      <td>599</td>\n",
       "      <td>993</td>\n",
       "      <td>2016</td>\n",
       "      <td>1433</td>\n",
       "      <td>3008</td>\n",
       "      <td>1979</td>\n",
       "      <td>2067</td>\n",
       "      <td>4900</td>\n",
       "      <td>2027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_4</th>\n",
       "      <td>5243</td>\n",
       "      <td>3526</td>\n",
       "      <td>4658</td>\n",
       "      <td>2475</td>\n",
       "      <td>5861</td>\n",
       "      <td>5526</td>\n",
       "      <td>3808</td>\n",
       "      <td>4545</td>\n",
       "      <td>4121</td>\n",
       "      <td>4551</td>\n",
       "      <td>...</td>\n",
       "      <td>7179</td>\n",
       "      <td>6226</td>\n",
       "      <td>10400</td>\n",
       "      <td>21570</td>\n",
       "      <td>21360</td>\n",
       "      <td>27375</td>\n",
       "      <td>29348</td>\n",
       "      <td>27454</td>\n",
       "      <td>26958</td>\n",
       "      <td>15939</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_5</th>\n",
       "      <td>3809</td>\n",
       "      <td>1128</td>\n",
       "      <td>1313</td>\n",
       "      <td>1989</td>\n",
       "      <td>3024</td>\n",
       "      <td>2419</td>\n",
       "      <td>1540</td>\n",
       "      <td>1956</td>\n",
       "      <td>869</td>\n",
       "      <td>2546</td>\n",
       "      <td>...</td>\n",
       "      <td>923</td>\n",
       "      <td>1287</td>\n",
       "      <td>2062</td>\n",
       "      <td>3568</td>\n",
       "      <td>2621</td>\n",
       "      <td>7346</td>\n",
       "      <td>3789</td>\n",
       "      <td>4428</td>\n",
       "      <td>9243</td>\n",
       "      <td>2211</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_639</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_640</th>\n",
       "      <td>4270</td>\n",
       "      <td>2662</td>\n",
       "      <td>3657</td>\n",
       "      <td>2136</td>\n",
       "      <td>4412</td>\n",
       "      <td>2622</td>\n",
       "      <td>1964</td>\n",
       "      <td>2673</td>\n",
       "      <td>1856</td>\n",
       "      <td>4256</td>\n",
       "      <td>...</td>\n",
       "      <td>312</td>\n",
       "      <td>248</td>\n",
       "      <td>302</td>\n",
       "      <td>535</td>\n",
       "      <td>498</td>\n",
       "      <td>755</td>\n",
       "      <td>607</td>\n",
       "      <td>602</td>\n",
       "      <td>1442</td>\n",
       "      <td>684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_641</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_642</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_643</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>643 rows × 86 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            sC0C  s1C3A  s1C3B  s1C5A  s1C5B  s1C7A  s1C7B  s1C10A  s1C10B  \\\n",
       "featureID                                                                    \n",
       "OTU_1          6     11     13     30     15     23     25       5       8   \n",
       "OTU_2      17546  18302  32674  15446  22216  25011  21224   13067   13340   \n",
       "OTU_3       3169   3033   4495   3138   4601   3633   1919    2952    2110   \n",
       "OTU_4       5243   3526   4658   2475   5861   5526   3808    4545    4121   \n",
       "OTU_5       3809   1128   1313   1989   3024   2419   1540    1956     869   \n",
       "...          ...    ...    ...    ...    ...    ...    ...     ...     ...   \n",
       "OTU_639        1      0      0      1      1      0      0       0       1   \n",
       "OTU_640     4270   2662   3657   2136   4412   2622   1964    2673    1856   \n",
       "OTU_641        0      0      0      1      3      3      0       5       0   \n",
       "OTU_642        0      0      0      0      0      0      0       0       0   \n",
       "OTU_643        0      0      0      0      0      0      0       0       0   \n",
       "\n",
       "           s1C20A  ...  s3G3B  s3G3C  s3G5A  s3G5B  s3G5C  s3G7A  s3G7B  \\\n",
       "featureID          ...                                                    \n",
       "OTU_1           0  ...     30     46      1     14     10      6      3   \n",
       "OTU_2        6896  ...   6869   8322   9792  21106  39348  24470  34872   \n",
       "OTU_3        5739  ...    610    599    993   2016   1433   3008   1979   \n",
       "OTU_4        4551  ...   7179   6226  10400  21570  21360  27375  29348   \n",
       "OTU_5        2546  ...    923   1287   2062   3568   2621   7346   3789   \n",
       "...           ...  ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "OTU_639         2  ...      0      0      0      0      0      0      0   \n",
       "OTU_640      4256  ...    312    248    302    535    498    755    607   \n",
       "OTU_641         0  ...      0      3      1      3      0      9      4   \n",
       "OTU_642         0  ...      0      0      0      0      0      0      0   \n",
       "OTU_643         1  ...      0      0      0      0      0      0      0   \n",
       "\n",
       "           s3G7C  s3G10A  s3G10B  \n",
       "featureID                         \n",
       "OTU_1          1      16      16  \n",
       "OTU_2      33561   16779   15614  \n",
       "OTU_3       2067    4900    2027  \n",
       "OTU_4      27454   26958   15939  \n",
       "OTU_5       4428    9243    2211  \n",
       "...          ...     ...     ...  \n",
       "OTU_639        0       0       0  \n",
       "OTU_640      602    1442     684  \n",
       "OTU_641        6       4       1  \n",
       "OTU_642        0       0       0  \n",
       "OTU_643        0       0       0  \n",
       "\n",
       "[643 rows x 86 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "OTU_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "OTU_counts.to_csv(\"../data/processed/OTU_counts.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create `feature_metadata.tsv` "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "OTU: OTU_18 Extra PART: species group  pseudomonas fluorescens group\n",
      "OTU: OTU_526 Extra PART: jpl_saf&apos\n",
      "OTU: OTU_526 Extra PART:  during msl mission spacecraft assembly clean room floor clone gi8_sp_i21  \n"
     ]
    }
   ],
   "source": [
    "# format table of OTU ID x taxonomy & function\n",
    "feature_metadata = pd.DataFrame(index=otu_raw.index, columns=[\"Name\"])\n",
    "\n",
    "for otu, desc in otu_raw.iterrows(): \n",
    "    parts = desc[\"16rRNA\"].split(\";\")\n",
    "    taxonomy = \"Unassigned\"\n",
    "    for i, part in enumerate(parts):\n",
    "        if i == 0:\n",
    "            feature_metadata.loc[otu, \"Name\"] = part\n",
    "        elif part.startswith(\"k__\"):\n",
    "            feature_metadata.loc[otu, \"kingdom\"] = part[3:]\n",
    "            taxonomy = \"Root;\" + part + \";\"\n",
    "        elif part.startswith(\"p__\"):\n",
    "            feature_metadata.loc[otu, \"phylum\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"c__\"):\n",
    "            feature_metadata.loc[otu, \"class\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"o__\"):\n",
    "            feature_metadata.loc[otu, \"order\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"f__\"):\n",
    "            feature_metadata.loc[otu, \"family\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"g__\"):\n",
    "            feature_metadata.loc[otu, \"genus\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"s__\"):\n",
    "            feature_metadata.loc[otu, \"species\"] = part[3:]\n",
    "            taxonomy += part + \";\"\n",
    "        elif part.startswith(\"superkingdom__\"):\n",
    "            feature_metadata.loc[otu, \"superkingdom\"] = part[14:]\n",
    "        elif part.startswith(\"superphylum__\"):\n",
    "            feature_metadata.loc[otu, \"superphylum\"] = part[13:]\n",
    "        elif part.startswith(\"subclass\"):\n",
    "            feature_metadata.loc[otu, \"subclass\"] = part[8:]\n",
    "        elif part.startswith(\"suborder\"):\n",
    "            feature_metadata.loc[otu, \"suborder\"] = part[8:]\n",
    "        elif part.startswith(\"subphylum__\"):\n",
    "            feature_metadata.loc[otu, \"subphylum\"] = part[9:]\n",
    "        else:\n",
    "            print(\"OTU:\", otu, \"Extra PART:\", part)\n",
    "    feature_metadata.loc[otu, \"taxonomy\"] = taxonomy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>kingdom</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>superkingdom</th>\n",
       "      <th>taxonomy</th>\n",
       "      <th>superphylum</th>\n",
       "      <th>subclass</th>\n",
       "      <th>suborder</th>\n",
       "      <th>subphylum</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>featureID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OTU_1</th>\n",
       "      <td>eu528231.1 seasonal dynamics mudflat mouth maj...</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>alphaproteobacteria</td>\n",
       "      <td>rhodospirillales</td>\n",
       "      <td>rhodospirillaceae</td>\n",
       "      <td>magnetospirillum</td>\n",
       "      <td>magnetospirillum sp.</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__proteobacteria;c__alphapro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_2</th>\n",
       "      <td>uncultured opitutus sp.</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>verrucomicrobia</td>\n",
       "      <td>opitutae</td>\n",
       "      <td>opitutales</td>\n",
       "      <td>opitutaceae</td>\n",
       "      <td>opitutus</td>\n",
       "      <td>opitutus spp.</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__verrucomicrobia;c__opituta...</td>\n",
       "      <td>chlamydiae/verrucomicrobia group</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_3</th>\n",
       "      <td>rhodopseudomonas palustris</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>alphaproteobacteria</td>\n",
       "      <td>rhizobiales</td>\n",
       "      <td>bradyrhizobiaceae</td>\n",
       "      <td>rhodopseudomonas</td>\n",
       "      <td>rhodopseudomonas palustris</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__proteobacteria;c__alphapro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_4</th>\n",
       "      <td>ay297802.1 waterlogged archaeological wood clo...</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>alphaproteobacteria</td>\n",
       "      <td>rhizobiales</td>\n",
       "      <td>methylocystaceae</td>\n",
       "      <td>pleomorphomonas</td>\n",
       "      <td>pleomorphomonas oryzae</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__proteobacteria;c__alphapro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_5</th>\n",
       "      <td>rhodoblastus acidophilus</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>alphaproteobacteria</td>\n",
       "      <td>rhizobiales</td>\n",
       "      <td>bradyrhizobiaceae</td>\n",
       "      <td>rhodoblastus</td>\n",
       "      <td>rhodoblastus acidophilus</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__proteobacteria;c__alphapro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_639</th>\n",
       "      <td>uncultured spirochaeta sp.</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>spirochaetes</td>\n",
       "      <td>spirochaetia</td>\n",
       "      <td>spirochaetales</td>\n",
       "      <td>spirochaetaceae</td>\n",
       "      <td>spirochaeta</td>\n",
       "      <td>spirochaeta spp.</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__spirochaetes;c__spirochaet...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_640</th>\n",
       "      <td>ab079679.1 rhodopseudomonas sp. hmd89</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>alphaproteobacteria</td>\n",
       "      <td>rhizobiales</td>\n",
       "      <td>bradyrhizobiaceae</td>\n",
       "      <td>afipia</td>\n",
       "      <td>afipia sp.</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__proteobacteria;c__alphapro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_641</th>\n",
       "      <td>ay297803.1 waterlogged archaeological wood clo...</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>alphaproteobacteria</td>\n",
       "      <td>rhizobiales</td>\n",
       "      <td>methylocystaceae</td>\n",
       "      <td>pleomorphomonas</td>\n",
       "      <td>pleomorphomonas oryzae</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__proteobacteria;c__alphapro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_642</th>\n",
       "      <td>uncultured pontibacter sp.</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>bacteroidetes</td>\n",
       "      <td>cytophagia</td>\n",
       "      <td>cytophagales</td>\n",
       "      <td>cytophagaceae</td>\n",
       "      <td>pontibacter</td>\n",
       "      <td>pontibacter spp.</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__bacteroidetes;c__cytophagi...</td>\n",
       "      <td>bacteroidetes/chlorobi group</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OTU_643</th>\n",
       "      <td>ay297802.1 waterlogged archaeological wood clo...</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>proteobacteria</td>\n",
       "      <td>alphaproteobacteria</td>\n",
       "      <td>rhizobiales</td>\n",
       "      <td>methylocystaceae</td>\n",
       "      <td>pleomorphomonas</td>\n",
       "      <td>pleomorphomonas oryzae</td>\n",
       "      <td>bacteria</td>\n",
       "      <td>Root;k__bacteria;p__proteobacteria;c__alphapro...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>643 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                        Name   kingdom  \\\n",
       "featureID                                                                \n",
       "OTU_1      eu528231.1 seasonal dynamics mudflat mouth maj...  bacteria   \n",
       "OTU_2                               uncultured opitutus sp.   bacteria   \n",
       "OTU_3                            rhodopseudomonas palustris   bacteria   \n",
       "OTU_4      ay297802.1 waterlogged archaeological wood clo...  bacteria   \n",
       "OTU_5                              rhodoblastus acidophilus   bacteria   \n",
       "...                                                      ...       ...   \n",
       "OTU_639                          uncultured spirochaeta sp.   bacteria   \n",
       "OTU_640              ab079679.1 rhodopseudomonas sp. hmd89    bacteria   \n",
       "OTU_641    ay297803.1 waterlogged archaeological wood clo...  bacteria   \n",
       "OTU_642                          uncultured pontibacter sp.   bacteria   \n",
       "OTU_643    ay297802.1 waterlogged archaeological wood clo...  bacteria   \n",
       "\n",
       "                    phylum                class             order  \\\n",
       "featureID                                                           \n",
       "OTU_1       proteobacteria  alphaproteobacteria  rhodospirillales   \n",
       "OTU_2      verrucomicrobia             opitutae        opitutales   \n",
       "OTU_3       proteobacteria  alphaproteobacteria       rhizobiales   \n",
       "OTU_4       proteobacteria  alphaproteobacteria       rhizobiales   \n",
       "OTU_5       proteobacteria  alphaproteobacteria       rhizobiales   \n",
       "...                    ...                  ...               ...   \n",
       "OTU_639       spirochaetes         spirochaetia    spirochaetales   \n",
       "OTU_640     proteobacteria  alphaproteobacteria       rhizobiales   \n",
       "OTU_641     proteobacteria  alphaproteobacteria       rhizobiales   \n",
       "OTU_642      bacteroidetes           cytophagia      cytophagales   \n",
       "OTU_643     proteobacteria  alphaproteobacteria       rhizobiales   \n",
       "\n",
       "                      family             genus                     species  \\\n",
       "featureID                                                                    \n",
       "OTU_1      rhodospirillaceae  magnetospirillum        magnetospirillum sp.   \n",
       "OTU_2            opitutaceae          opitutus               opitutus spp.   \n",
       "OTU_3      bradyrhizobiaceae  rhodopseudomonas  rhodopseudomonas palustris   \n",
       "OTU_4       methylocystaceae   pleomorphomonas      pleomorphomonas oryzae   \n",
       "OTU_5      bradyrhizobiaceae      rhodoblastus    rhodoblastus acidophilus   \n",
       "...                      ...               ...                         ...   \n",
       "OTU_639      spirochaetaceae       spirochaeta            spirochaeta spp.   \n",
       "OTU_640    bradyrhizobiaceae            afipia                  afipia sp.   \n",
       "OTU_641     methylocystaceae   pleomorphomonas      pleomorphomonas oryzae   \n",
       "OTU_642        cytophagaceae       pontibacter            pontibacter spp.   \n",
       "OTU_643     methylocystaceae   pleomorphomonas      pleomorphomonas oryzae   \n",
       "\n",
       "          superkingdom                                           taxonomy  \\\n",
       "featureID                                                                   \n",
       "OTU_1         bacteria  Root;k__bacteria;p__proteobacteria;c__alphapro...   \n",
       "OTU_2         bacteria  Root;k__bacteria;p__verrucomicrobia;c__opituta...   \n",
       "OTU_3         bacteria  Root;k__bacteria;p__proteobacteria;c__alphapro...   \n",
       "OTU_4         bacteria  Root;k__bacteria;p__proteobacteria;c__alphapro...   \n",
       "OTU_5         bacteria  Root;k__bacteria;p__proteobacteria;c__alphapro...   \n",
       "...                ...                                                ...   \n",
       "OTU_639       bacteria  Root;k__bacteria;p__spirochaetes;c__spirochaet...   \n",
       "OTU_640       bacteria  Root;k__bacteria;p__proteobacteria;c__alphapro...   \n",
       "OTU_641       bacteria  Root;k__bacteria;p__proteobacteria;c__alphapro...   \n",
       "OTU_642       bacteria  Root;k__bacteria;p__bacteroidetes;c__cytophagi...   \n",
       "OTU_643       bacteria  Root;k__bacteria;p__proteobacteria;c__alphapro...   \n",
       "\n",
       "                                superphylum subclass suborder subphylum  \n",
       "featureID                                                                \n",
       "OTU_1                                   NaN      NaN      NaN       NaN  \n",
       "OTU_2      chlamydiae/verrucomicrobia group      NaN      NaN       NaN  \n",
       "OTU_3                                   NaN      NaN      NaN       NaN  \n",
       "OTU_4                                   NaN      NaN      NaN       NaN  \n",
       "OTU_5                                   NaN      NaN      NaN       NaN  \n",
       "...                                     ...      ...      ...       ...  \n",
       "OTU_639                                 NaN      NaN      NaN       NaN  \n",
       "OTU_640                                 NaN      NaN      NaN       NaN  \n",
       "OTU_641                                 NaN      NaN      NaN       NaN  \n",
       "OTU_642        bacteroidetes/chlorobi group      NaN      NaN       NaN  \n",
       "OTU_643                                 NaN      NaN      NaN       NaN  \n",
       "\n",
       "[643 rows x 14 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "feature_metadata.to_csv(\"../data/processed/feature_metadata.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create `taxonomy.tsv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "featureID\n",
       "OTU_1      Root;k__bacteria;p__proteobacteria;c__alphapro...\n",
       "OTU_2      Root;k__bacteria;p__verrucomicrobia;c__opituta...\n",
       "OTU_3      Root;k__bacteria;p__proteobacteria;c__alphapro...\n",
       "OTU_4      Root;k__bacteria;p__proteobacteria;c__alphapro...\n",
       "OTU_5      Root;k__bacteria;p__proteobacteria;c__alphapro...\n",
       "                                 ...                        \n",
       "OTU_639    Root;k__bacteria;p__spirochaetes;c__spirochaet...\n",
       "OTU_640    Root;k__bacteria;p__proteobacteria;c__alphapro...\n",
       "OTU_641    Root;k__bacteria;p__proteobacteria;c__alphapro...\n",
       "OTU_642    Root;k__bacteria;p__bacteroidetes;c__cytophagi...\n",
       "OTU_643    Root;k__bacteria;p__proteobacteria;c__alphapro...\n",
       "Name: taxonomy, Length: 643, dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# save\n",
    "feature_metadata[\"taxonomy\"].to_csv(\"../data/processed/taxonomy.tsv\", sep='\\t')\n",
    "feature_metadata[\"taxonomy\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Additional Cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sC0C</th>\n",
       "      <th>s1C3A</th>\n",
       "      <th>s1C3B</th>\n",
       "      <th>s1C5A</th>\n",
       "      <th>s1C5B</th>\n",
       "      <th>s1C7A</th>\n",
       "      <th>s1C7B</th>\n",
       "      <th>s1C10A</th>\n",
       "      <th>s1C10B</th>\n",
       "      <th>s1C20A</th>\n",
       "      <th>...</th>\n",
       "      <th>s3G3B</th>\n",
       "      <th>s3G3C</th>\n",
       "      <th>s3G5A</th>\n",
       "      <th>s3G5B</th>\n",
       "      <th>s3G5C</th>\n",
       "      <th>s3G7A</th>\n",
       "      <th>s3G7B</th>\n",
       "      <th>s3G7C</th>\n",
       "      <th>s3G10A</th>\n",
       "      <th>s3G10B</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>featureID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>0 rows × 86 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [sC0C, s1C3A, s1C3B, s1C5A, s1C5B, s1C7A, s1C7B, s1C10A, s1C10B, s1C20A, s1C20B, s2C3A, s2C3B, s2C5A, s2C5B, s2C10A, s2C10B, s1M3A, s1M3B, s1M3C, s1M5C, s1M5A, s1M5B, s1M7A, s1M7B, s1M7C, s1M10A, s1M10B, s1M10C, s2M3A, s2M3B, s2M3C, s2M5A, s2M5B, s2M7A, s2M7B, s2M10B, s2M10A, s2M10C, s3M3A, s3M3B, s3M3C, s3M5A, s3M5B, s3M5C, s3M7A, s3M7B, s3M7C, s3M10A, s3M10B, s3M10C, s1G3A, s1G3B, s1G3C, s1G5A, s1G5B, s1G5C, s1G7A, s1G7B, s1G7C, s1G10A, s1G10B, s1G10C, s2G3A, s2G3B, s2G3C, s2G5A, s2G5B, s2G5C, s2G7A, s2G7B, s2G7C, s2G10A, s2G10B, s2G10C, s3G3A, s3G3B, s3G3C, s3G5A, s3G5B, s3G5C, s3G7A, s3G7B, s3G7C, s3G10A, s3G10B]\n",
       "Index: []\n",
       "\n",
       "[0 rows x 86 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check if there are OTUs that appear in no samples\n",
    "OTU_counts[OTU_counts.sum(axis=1) == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "featureID\n",
      "OTU_55     uncultured methanosaeta sp. \n",
      "OTU_97                    No hits found\n",
      "OTU_112                   No hits found\n",
      "OTU_124                   No hits found\n",
      "OTU_152                   No hits found\n",
      "OTU_154                   No hits found\n",
      "OTU_221                 vitis vinifera \n",
      "Name: Name, dtype: object\n",
      "OTU_counts.shape (643, 86) OTU_counts_clean.shape (636, 86)\n",
      "feature_metadata.shape (643, 14) feature_metadata_clean.shape (636, 14)\n"
     ]
    }
   ],
   "source": [
    "# remove OTUs that are not bacteria\n",
    "not_bacteria = feature_metadata[feature_metadata[\"kingdom\"] != \"bacteria\"]\n",
    "print(not_bacteria.Name)\n",
    "\n",
    "OTU_counts_clean = OTU_counts.drop(not_bacteria.index)\n",
    "print(\"OTU_counts.shape\", OTU_counts.shape, \"OTU_counts_clean.shape\", OTU_counts_clean.shape,)\n",
    "\n",
    "feature_metadata_clean = feature_metadata.drop(not_bacteria.index)\n",
    "print(\"feature_metadata.shape\", feature_metadata.shape, \"feature_metadata_clean.shape\", feature_metadata_clean.shape,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "featureID\n",
      "OTU_19     74\n",
      "OTU_31     62\n",
      "OTU_38     29\n",
      "OTU_66     18\n",
      "OTU_86     70\n",
      "           ..\n",
      "OTU_637     4\n",
      "OTU_638    18\n",
      "OTU_639     9\n",
      "OTU_642     5\n",
      "OTU_643    18\n",
      "Length: 406, dtype: int64\n",
      "OTU_counts_clean.shape (230, 86)\n",
      "feature_metadata_clean.shape (230, 14)\n"
     ]
    }
   ],
   "source": [
    "# remove OTUs with less than threshold total reads\n",
    "threshold_OTU_total_reads = 100\n",
    "less_than_threshold = OTU_counts_clean[OTU_counts_clean.sum(axis=1) < threshold_OTU_total_reads]\n",
    "print(less_than_threshold.sum(axis=1))\n",
    "\n",
    "OTU_counts_clean = OTU_counts_clean.drop(less_than_threshold.index)\n",
    "print(\"OTU_counts_clean.shape\", OTU_counts_clean.shape,)\n",
    "\n",
    "feature_metadata_clean = feature_metadata_clean.drop(less_than_threshold.index)\n",
    "print(\"feature_metadata_clean.shape\", feature_metadata_clean.shape,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save clean\n",
    "OTU_counts_clean.to_csv(\"../data/processed/OTU_counts_clean.tsv\", sep='\\t')\n",
    "feature_metadata_clean.to_csv(\"../data/processed/feature_metadata_clean.tsv\", sep='\\t')\n",
    "feature_metadata_clean[[\"kingdom\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"]].to_csv(\"../data/processed/taxonomy_table_clean.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get Phylogenetic Info and Alignment Scores\n",
    "\n",
    "* OTU ID: DNA Sequences\n",
    "    * (MR DNA): `081616JHnew515Fcomplete-pr.fasta.otus.fa`\n",
    "* OTU ID x OTU ID Pairwise Alignment Scores\n",
    "    * (Muscle): `Pairwise_distances_all_OTU_Muscle`\n",
    "* OTU ID Phylogenetic Tree\n",
    "    * (Muscle): `MUSCLE_alignment_ML_tree_all_OTUs`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Rename OTU ID: DNA Sequences mapping (for Jean, to visually analyze phylogenetic tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from skbio.sequence import DNA\n",
    "\n",
    "sequence_file = open(\"../data/raw/081616JHnew515Fcomplete-pr.fasta.otus.fa\", 'r')\n",
    "sequence_lines = [l.upper() for l in sequence_file.readlines()]\n",
    "dna_lines = [DNA(str(seq), metadata=seq.metadata) for seq in skbio.io.read(sequence_lines, format='fasta')]\n",
    "dna_lines_renamed = []\n",
    "\n",
    "for seq in skbio.io.read(sequence_lines, format='fasta'):\n",
    "    str_seq = str(seq)\n",
    "    metadata = seq.metadata\n",
    "    metadata['id'] = metadata['id'] + \"_\" + feature_metadata.loc[metadata['id'], \"Name\"].replace(\" \", \"_\").replace(\".\", \"\")\n",
    "    dna_lines_renamed.append(DNA(str_seq, metadata=metadata))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save\n",
    "dna_lines_renamed_file = open(\"../data/processed/renamed_081616JHnew515Fcomplete-pr.fasta.otus.fa\", \"w\")\n",
    "for seq in dna_lines_renamed:\n",
    "    skbio.io.write(seq, into=dna_lines_renamed_file, format='fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import and Use Newick Phylogenetic Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(((((((((((((((((((((((((((((((((((((((OTU_199:0.0 ...\n"
     ]
    }
   ],
   "source": [
    "# import phylogenetic weights and use in an example\n",
    "from skbio import TreeNode\n",
    "tree_file = open(\"../data/raw/MUSCLE_alignment_ML_tree_all_OTUs\", \"r\")\n",
    "tree = TreeNode.read(tree_file)\n",
    "tree_rooted = tree.root_at_midpoint()\n",
    "print(str(tree)[:50], \"...\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sC0C      17.636550\n",
       "s1C3A     15.217088\n",
       "s1C3B     17.104466\n",
       "s1C5A     16.736714\n",
       "s1C5B     19.248967\n",
       "            ...    \n",
       "s3G7A     20.697255\n",
       "s3G7B     20.340444\n",
       "s3G7C     19.767499\n",
       "s3G10A    20.267999\n",
       "s3G10B    19.147051\n",
       "Length: 86, dtype: float64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# alpha diversity w/ phylogenetic weights example \n",
    "from skbio.diversity import alpha_diversity\n",
    "alpha_diversity(metric=\"faith_pd\", counts=OTU_counts.T.values, ids=OTU_counts.columns, tree=tree_rooted, otu_ids=[n.name for n in tree_rooted.tips()], validate=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experiment Scratch-Pad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
